{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "7100e30d",
   "metadata": {},
   "source": [
    "## Library and data import"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c5e03f94",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Attaching package: ‘rstatix’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:stats’:\n",
      "\n",
      "    filter\n",
      "\n",
      "\n",
      "── \u001b[1mAttaching core tidyverse packages\u001b[22m ──────────────────────── tidyverse 2.0.0 ──\n",
      "\u001b[32m✔\u001b[39m \u001b[34mdplyr    \u001b[39m 1.1.4     \u001b[32m✔\u001b[39m \u001b[34mpurrr    \u001b[39m 1.1.0\n",
      "\u001b[32m✔\u001b[39m \u001b[34mforcats  \u001b[39m 1.0.1     \u001b[32m✔\u001b[39m \u001b[34mreadr    \u001b[39m 2.1.5\n",
      "\u001b[32m✔\u001b[39m \u001b[34mggplot2  \u001b[39m 4.0.0     \u001b[32m✔\u001b[39m \u001b[34mtibble   \u001b[39m 3.3.0\n",
      "\u001b[32m✔\u001b[39m \u001b[34mlubridate\u001b[39m 1.9.4     \u001b[32m✔\u001b[39m \u001b[34mtidyr    \u001b[39m 1.3.1\n",
      "── \u001b[1mConflicts\u001b[22m ────────────────────────────────────────── tidyverse_conflicts() ──\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfilter()\u001b[39m masks \u001b[34mrstatix\u001b[39m::filter(), \u001b[34mstats\u001b[39m::filter()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mlag()\u001b[39m    masks \u001b[34mstats\u001b[39m::lag()\n",
      "\u001b[36mℹ\u001b[39m Use the conflicted package (\u001b[3m\u001b[34m<http://conflicted.r-lib.org/>\u001b[39m\u001b[23m) to force all conflicts to become errors\n"
     ]
    }
   ],
   "source": [
    "library(rstatix)\n",
    "library(stringr)\n",
    "library(tidyverse)\n",
    "library(RColorBrewer)\n",
    "library(dplyr)\n",
    "library(tidyr)\n",
    "library(tibble)\n",
    "library(readr)\n",
    "library(ggplot2)\n",
    "library(fs)\n",
    "library(gt)\n",
    "library(readr)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d82447ee",
   "metadata": {},
   "source": [
    "STRs data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "030e72bc",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1mRows: \u001b[22m\u001b[34m3333995\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m19\u001b[39m\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \"\\t\"\n",
      "\u001b[31mchr\u001b[39m  (9): STRs_ID, region, gene_id, gene_name, gene_chrom, annotation, sampl...\n",
      "\u001b[32mdbl\u001b[39m (10): priority, gene_start, gene_end, start, end, allele1_est, allele2_e...\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 6 × 19</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>STRs_ID</th><th scope=col>region</th><th scope=col>gene_id</th><th scope=col>priority</th><th scope=col>gene_name</th><th scope=col>gene_chrom</th><th scope=col>gene_start</th><th scope=col>gene_end</th><th scope=col>annotation</th><th scope=col>sample_id</th><th scope=col>chrom</th><th scope=col>start</th><th scope=col>end</th><th scope=col>repeat_unit</th><th scope=col>allele1_est</th><th scope=col>allele2_est</th><th scope=col>depth</th><th scope=col>p</th><th scope=col>p_adj</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>chr18:35064000:A:34</td><td>intron    </td><td>ENSG00000166974</td><td>5</td><td>MAPRE2 </td><td>chr18</td><td>34976928</td><td>35143470</td><td>intron:MAPRE2 </td><td>BEL-1012-1</td><td>18</td><td> 35064000</td><td> 35064033</td><td>A </td><td>-9</td><td>155.68</td><td>33</td><td>1.7e-157</td><td>1.5e-152</td></tr>\n",
       "\t<tr><td>chr1:69249900:T:2  </td><td>intergenic</td><td>intergenic     </td><td>6</td><td>.      </td><td>.    </td><td>      -1</td><td>      -1</td><td>intergenic    </td><td>BEL-1012-1</td><td>1 </td><td> 69249900</td><td> 69249901</td><td>T </td><td> 0</td><td>112.16</td><td>32</td><td>8.8e-143</td><td>3.9e-138</td></tr>\n",
       "\t<tr><td>chr4:93128027:A:33 </td><td>intron    </td><td>ENSG00000152208</td><td>5</td><td>GRID2  </td><td>chr4 </td><td>92303966</td><td>93810157</td><td>intron:GRID2  </td><td>BEL-1012-1</td><td>4 </td><td> 93128027</td><td> 93128059</td><td>A </td><td> 0</td><td>151.43</td><td>37</td><td>2.5e-127</td><td>2.0e-123</td></tr>\n",
       "\t<tr><td>chr1:108526008:A:2 </td><td>intergenic</td><td>intergenic     </td><td>6</td><td>.      </td><td>.    </td><td>      -1</td><td>      -1</td><td>intergenic    </td><td>BEL-1012-1</td><td>1 </td><td>108526008</td><td>108526009</td><td>A </td><td> 0</td><td> 77.30</td><td>38</td><td> 1.9e-88</td><td> 1.6e-85</td></tr>\n",
       "\t<tr><td>chr12:46783052:GT:4</td><td>intron    </td><td>ENSG00000139209</td><td>5</td><td>SLC38A4</td><td>chr12</td><td>46764761</td><td>46832408</td><td>intron:SLC38A4</td><td>BEL-1012-1</td><td>12</td><td> 46783052</td><td> 46783059</td><td>GT</td><td> 0</td><td> 19.24</td><td>33</td><td> 1.1e-46</td><td> 9.8e-45</td></tr>\n",
       "\t<tr><td>chr11:14508561:A:17</td><td>intron    </td><td>ENSG00000129084</td><td>5</td><td>PSMA1  </td><td>chr11</td><td>14504874</td><td>14643635</td><td>intron:PSMA1  </td><td>BEL-1012-1</td><td>11</td><td> 14508561</td><td> 14508577</td><td>A </td><td> 0</td><td> 88.74</td><td>38</td><td> 1.1e-05</td><td> 9.6e-05</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 6 × 19\n",
       "\\begin{tabular}{lllllllllllllllllll}\n",
       " STRs\\_ID & region & gene\\_id & priority & gene\\_name & gene\\_chrom & gene\\_start & gene\\_end & annotation & sample\\_id & chrom & start & end & repeat\\_unit & allele1\\_est & allele2\\_est & depth & p & p\\_adj\\\\\n",
       " <chr> & <chr> & <chr> & <dbl> & <chr> & <chr> & <dbl> & <dbl> & <chr> & <chr> & <chr> & <dbl> & <dbl> & <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t chr18:35064000:A:34 & intron     & ENSG00000166974 & 5 & MAPRE2  & chr18 & 34976928 & 35143470 & intron:MAPRE2  & BEL-1012-1 & 18 &  35064000 &  35064033 & A  & -9 & 155.68 & 33 & 1.7e-157 & 1.5e-152\\\\\n",
       "\t chr1:69249900:T:2   & intergenic & intergenic      & 6 & .       & .     &       -1 &       -1 & intergenic     & BEL-1012-1 & 1  &  69249900 &  69249901 & T  &  0 & 112.16 & 32 & 8.8e-143 & 3.9e-138\\\\\n",
       "\t chr4:93128027:A:33  & intron     & ENSG00000152208 & 5 & GRID2   & chr4  & 92303966 & 93810157 & intron:GRID2   & BEL-1012-1 & 4  &  93128027 &  93128059 & A  &  0 & 151.43 & 37 & 2.5e-127 & 2.0e-123\\\\\n",
       "\t chr1:108526008:A:2  & intergenic & intergenic      & 6 & .       & .     &       -1 &       -1 & intergenic     & BEL-1012-1 & 1  & 108526008 & 108526009 & A  &  0 &  77.30 & 38 &  1.9e-88 &  1.6e-85\\\\\n",
       "\t chr12:46783052:GT:4 & intron     & ENSG00000139209 & 5 & SLC38A4 & chr12 & 46764761 & 46832408 & intron:SLC38A4 & BEL-1012-1 & 12 &  46783052 &  46783059 & GT &  0 &  19.24 & 33 &  1.1e-46 &  9.8e-45\\\\\n",
       "\t chr11:14508561:A:17 & intron     & ENSG00000129084 & 5 & PSMA1   & chr11 & 14504874 & 14643635 & intron:PSMA1   & BEL-1012-1 & 11 &  14508561 &  14508577 & A  &  0 &  88.74 & 38 &  1.1e-05 &  9.6e-05\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 6 × 19\n",
       "\n",
       "| STRs_ID &lt;chr&gt; | region &lt;chr&gt; | gene_id &lt;chr&gt; | priority &lt;dbl&gt; | gene_name &lt;chr&gt; | gene_chrom &lt;chr&gt; | gene_start &lt;dbl&gt; | gene_end &lt;dbl&gt; | annotation &lt;chr&gt; | sample_id &lt;chr&gt; | chrom &lt;chr&gt; | start &lt;dbl&gt; | end &lt;dbl&gt; | repeat_unit &lt;chr&gt; | allele1_est &lt;dbl&gt; | allele2_est &lt;dbl&gt; | depth &lt;dbl&gt; | p &lt;dbl&gt; | p_adj &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| chr18:35064000:A:34 | intron     | ENSG00000166974 | 5 | MAPRE2  | chr18 | 34976928 | 35143470 | intron:MAPRE2  | BEL-1012-1 | 18 |  35064000 |  35064033 | A  | -9 | 155.68 | 33 | 1.7e-157 | 1.5e-152 |\n",
       "| chr1:69249900:T:2   | intergenic | intergenic      | 6 | .       | .     |       -1 |       -1 | intergenic     | BEL-1012-1 | 1  |  69249900 |  69249901 | T  |  0 | 112.16 | 32 | 8.8e-143 | 3.9e-138 |\n",
       "| chr4:93128027:A:33  | intron     | ENSG00000152208 | 5 | GRID2   | chr4  | 92303966 | 93810157 | intron:GRID2   | BEL-1012-1 | 4  |  93128027 |  93128059 | A  |  0 | 151.43 | 37 | 2.5e-127 | 2.0e-123 |\n",
       "| chr1:108526008:A:2  | intergenic | intergenic      | 6 | .       | .     |       -1 |       -1 | intergenic     | BEL-1012-1 | 1  | 108526008 | 108526009 | A  |  0 |  77.30 | 38 |  1.9e-88 |  1.6e-85 |\n",
       "| chr12:46783052:GT:4 | intron     | ENSG00000139209 | 5 | SLC38A4 | chr12 | 46764761 | 46832408 | intron:SLC38A4 | BEL-1012-1 | 12 |  46783052 |  46783059 | GT |  0 |  19.24 | 33 |  1.1e-46 |  9.8e-45 |\n",
       "| chr11:14508561:A:17 | intron     | ENSG00000129084 | 5 | PSMA1   | chr11 | 14504874 | 14643635 | intron:PSMA1   | BEL-1012-1 | 11 |  14508561 |  14508577 | A  |  0 |  88.74 | 38 |  1.1e-05 |  9.6e-05 |\n",
       "\n"
      ],
      "text/plain": [
       "  STRs_ID             region     gene_id         priority gene_name gene_chrom\n",
       "1 chr18:35064000:A:34 intron     ENSG00000166974 5        MAPRE2    chr18     \n",
       "2 chr1:69249900:T:2   intergenic intergenic      6        .         .         \n",
       "3 chr4:93128027:A:33  intron     ENSG00000152208 5        GRID2     chr4      \n",
       "4 chr1:108526008:A:2  intergenic intergenic      6        .         .         \n",
       "5 chr12:46783052:GT:4 intron     ENSG00000139209 5        SLC38A4   chr12     \n",
       "6 chr11:14508561:A:17 intron     ENSG00000129084 5        PSMA1     chr11     \n",
       "  gene_start gene_end annotation     sample_id  chrom start     end      \n",
       "1 34976928   35143470 intron:MAPRE2  BEL-1012-1 18     35064000  35064033\n",
       "2       -1         -1 intergenic     BEL-1012-1 1      69249900  69249901\n",
       "3 92303966   93810157 intron:GRID2   BEL-1012-1 4      93128027  93128059\n",
       "4       -1         -1 intergenic     BEL-1012-1 1     108526008 108526009\n",
       "5 46764761   46832408 intron:SLC38A4 BEL-1012-1 12     46783052  46783059\n",
       "6 14504874   14643635 intron:PSMA1   BEL-1012-1 11     14508561  14508577\n",
       "  repeat_unit allele1_est allele2_est depth p        p_adj   \n",
       "1 A           -9          155.68      33    1.7e-157 1.5e-152\n",
       "2 T            0          112.16      32    8.8e-143 3.9e-138\n",
       "3 A            0          151.43      37    2.5e-127 2.0e-123\n",
       "4 A            0           77.30      38     1.9e-88  1.6e-85\n",
       "5 GT           0           19.24      33     1.1e-46  9.8e-45\n",
       "6 A            0           88.74      38     1.1e-05  9.6e-05"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'STRs_ID'</li><li>'region'</li><li>'gene_id'</li><li>'priority'</li><li>'gene_name'</li><li>'gene_chrom'</li><li>'gene_start'</li><li>'gene_end'</li><li>'annotation'</li><li>'sample_id'</li><li>'chrom'</li><li>'start'</li><li>'end'</li><li>'repeat_unit'</li><li>'allele1_est'</li><li>'allele2_est'</li><li>'depth'</li><li>'p'</li><li>'p_adj'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'STRs\\_ID'\n",
       "\\item 'region'\n",
       "\\item 'gene\\_id'\n",
       "\\item 'priority'\n",
       "\\item 'gene\\_name'\n",
       "\\item 'gene\\_chrom'\n",
       "\\item 'gene\\_start'\n",
       "\\item 'gene\\_end'\n",
       "\\item 'annotation'\n",
       "\\item 'sample\\_id'\n",
       "\\item 'chrom'\n",
       "\\item 'start'\n",
       "\\item 'end'\n",
       "\\item 'repeat\\_unit'\n",
       "\\item 'allele1\\_est'\n",
       "\\item 'allele2\\_est'\n",
       "\\item 'depth'\n",
       "\\item 'p'\n",
       "\\item 'p\\_adj'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'STRs_ID'\n",
       "2. 'region'\n",
       "3. 'gene_id'\n",
       "4. 'priority'\n",
       "5. 'gene_name'\n",
       "6. 'gene_chrom'\n",
       "7. 'gene_start'\n",
       "8. 'gene_end'\n",
       "9. 'annotation'\n",
       "10. 'sample_id'\n",
       "11. 'chrom'\n",
       "12. 'start'\n",
       "13. 'end'\n",
       "14. 'repeat_unit'\n",
       "15. 'allele1_est'\n",
       "16. 'allele2_est'\n",
       "17. 'depth'\n",
       "18. 'p'\n",
       "19. 'p_adj'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"STRs_ID\"     \"region\"      \"gene_id\"     \"priority\"    \"gene_name\"  \n",
       " [6] \"gene_chrom\"  \"gene_start\"  \"gene_end\"    \"annotation\"  \"sample_id\"  \n",
       "[11] \"chrom\"       \"start\"       \"end\"         \"repeat_unit\" \"allele1_est\"\n",
       "[16] \"allele2_est\" \"depth\"       \"p\"           \"p_adj\"      "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "df_strs <- read_tsv(\"../samples/gp_global/merged/outliers_annotated_complete.tsv\")\n",
    "\n",
    "head(df_strs)\n",
    "colnames(df_strs)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d3324c46",
   "metadata": {},
   "source": [
    "Import list of genes expressed in lung tissue (COVID-19), identified via sc-RNAseq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c0138364",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1mRows: \u001b[22m\u001b[34m2154\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m8\u001b[39m\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \",\"\n",
      "\u001b[31mchr\u001b[39m (6): Gene Symbol, Ensembl ID, Dataset, Accession, Tissue, origem\n",
      "\u001b[32mdbl\u001b[39m (2): Pvalue, LogFC\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "--- DEBUG: Initial Counts ---\n",
      "Total STRs (observations): 3333995\n",
      "Total Gene Entries (context rows): 2154\n",
      "\n",
      "--- DEBUG: Merge Results ---\n",
      "Final Rows (Expanded): 341923\n",
      "Unique STR Observations covered: 144796\n",
      "Note: The row count reflects the expansion where each STR is listed for every tissue/dataset context of its gene.\n",
      "\n",
      "--- Preview ---\n",
      "\u001b[90m# A tibble: 6 × 25\u001b[39m\n",
      "  STRs_ID       region gene_id priority gene_name gene_chrom gene_start gene_end\n",
      "  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m         \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m      \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m     \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m           \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m    \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m\n",
      "\u001b[90m1\u001b[39m chr17:667995… intron ENSG00…        5 PRKCA     chr17        66\u001b[4m3\u001b[24m\u001b[4m0\u001b[24m\u001b[4m2\u001b[24m613   6.68\u001b[90me\u001b[39m7\n",
      "\u001b[90m2\u001b[39m chr17:667995… intron ENSG00…        5 PRKCA     chr17        66\u001b[4m3\u001b[24m\u001b[4m0\u001b[24m\u001b[4m2\u001b[24m613   6.68\u001b[90me\u001b[39m7\n",
      "\u001b[90m3\u001b[39m chr7:1820753… intron ENSG00…        5 HDAC9     chr7         18\u001b[4m0\u001b[24m\u001b[4m8\u001b[24m\u001b[4m6\u001b[24m949   1.90\u001b[90me\u001b[39m7\n",
      "\u001b[90m4\u001b[39m chr18:687319… intron ENSG00…        5 CCDC102B  chr18        68\u001b[4m7\u001b[24m\u001b[4m1\u001b[24m\u001b[4m5\u001b[24m209   6.91\u001b[90me\u001b[39m7\n",
      "\u001b[90m5\u001b[39m chr3:1058690… intron ENSG00…        5 CBLB      chr3        105\u001b[4m6\u001b[24m\u001b[4m5\u001b[24m\u001b[4m5\u001b[24m461   1.06\u001b[90me\u001b[39m8\n",
      "\u001b[90m6\u001b[39m chr6:1292224… intron ENSG00…        5 LAMA2     chr6        128\u001b[4m8\u001b[24m\u001b[4m8\u001b[24m\u001b[4m3\u001b[24m141   1.30\u001b[90me\u001b[39m8\n",
      "\u001b[90m# ℹ 17 more variables: annotation <chr>, sample_id <chr>, chrom <chr>,\u001b[39m\n",
      "\u001b[90m#   start <dbl>, end <dbl>, repeat_unit <chr>, allele1_est <dbl>,\u001b[39m\n",
      "\u001b[90m#   allele2_est <dbl>, depth <dbl>, p <dbl>, p_adj <dbl>, `Ensembl ID` <chr>,\u001b[39m\n",
      "\u001b[90m#   Pvalue <dbl>, LogFC <dbl>, Dataset <chr>, Tissue <chr>, origem <chr>\u001b[39m\n"
     ]
    }
   ],
   "source": [
    "# 1. Load Gene Data\n",
    "# We do NOT remove gene duplicates here, as we want all entries\n",
    "df_genes <- read_csv(\"merged_scovid.csv\") %>%\n",
    "  select(-Accession) # Remove only irrelevant technical columns\n",
    "\n",
    "# 2. DEBUG: Initial Counts\n",
    "cat(\"\\n--- DEBUG: Initial Counts ---\\n\")\n",
    "n_total_strs <- nrow(df_strs) # Your original STR dataframe (multiple patients)\n",
    "n_gene_entries <- nrow(df_genes) # Entries in the gene list (may have repeated genes per tissue)\n",
    "\n",
    "cat(sprintf(\"Total STRs (observations): %d\\n\", n_total_strs))\n",
    "cat(sprintf(\"Total Gene Entries (context rows): %d\\n\", n_gene_entries))\n",
    "\n",
    "# 3. Full Merge (Many-to-Many)\n",
    "df_filtered <- df_strs %>%\n",
    "  inner_join(df_genes, by = c(\"gene_name\" = \"Gene Symbol\"), relationship = \"many-to-many\")\n",
    "\n",
    "# 4. DEBUG: Merge Results\n",
    "n_final_rows <- nrow(df_filtered)\n",
    "\n",
    "# Calculating how many unique STR observations (Locus + Patient) were actually kept\n",
    "n_unique_strs <- n_distinct(df_filtered$STRs_ID, df_filtered$sample_id) \n",
    "\n",
    "cat(\"\\n--- DEBUG: Merge Results ---\\n\")\n",
    "cat(sprintf(\"Final Rows (Expanded): %d\\n\", n_final_rows))\n",
    "cat(sprintf(\"Unique STR Observations covered: %d\\n\", n_unique_strs))\n",
    "cat(\"Note: The row count reflects the expansion where each STR is listed for every tissue/dataset context of its gene.\\n\")\n",
    "\n",
    "# 5. Preview\n",
    "cat(\"\\n--- Preview ---\\n\")\n",
    "print(head(df_filtered))\n",
    "\n",
    "# Save results\n",
    "output_dir <- \"./results\"\n",
    "  if (!dir.exists(output_dir)) dir.create(output_dir)\n",
    "  write_csv(df_filtered, file.path(output_dir, \"STRs_COVID_Expanded_Context.csv\"))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9370e820",
   "metadata": {},
   "source": [
    "Import association between samples and groups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "9cc171d2",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1mRows: \u001b[22m\u001b[34m168\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m2\u001b[39m\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \",\"\n",
      "\u001b[31mchr\u001b[39m (2): sample, group\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "--- DEBUG: Checking ID Match ---\n",
      "Total unique samples in STR data: 168\n",
      "Total samples with group info found: 168\n",
      "SUCCESS: All samples have group info.\n",
      "\n",
      "--- Preview with Groups ---\n",
      "\u001b[90m# A tibble: 6 × 26\u001b[39m\n",
      "  sample_id   group STRs_ID         region gene_id priority gene_name gene_chrom\n",
      "  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m       \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m           \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m      \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m     \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m     \n",
      "\u001b[90m1\u001b[39m GIG-RS-15-1 case  chr17:66799505… intron ENSG00…        5 PRKCA     chr17     \n",
      "\u001b[90m2\u001b[39m GIG-RS-15-1 case  chr17:66799505… intron ENSG00…        5 PRKCA     chr17     \n",
      "\u001b[90m3\u001b[39m GIG-RS-15-1 case  chr7:18207539:… intron ENSG00…        5 HDAC9     chr7      \n",
      "\u001b[90m4\u001b[39m GIG-RS-15-1 case  chr18:68731924… intron ENSG00…        5 CCDC102B  chr18     \n",
      "\u001b[90m5\u001b[39m GIG-RS-15-1 case  chr3:105869062… intron ENSG00…        5 CBLB      chr3      \n",
      "\u001b[90m6\u001b[39m GIG-RS-15-1 case  chr6:129222479… intron ENSG00…        5 LAMA2     chr6      \n",
      "\u001b[90m# ℹ 18 more variables: gene_start <dbl>, gene_end <dbl>, annotation <chr>,\u001b[39m\n",
      "\u001b[90m#   chrom <chr>, start <dbl>, end <dbl>, repeat_unit <chr>, allele1_est <dbl>,\u001b[39m\n",
      "\u001b[90m#   allele2_est <dbl>, depth <dbl>, p <dbl>, p_adj <dbl>, `Ensembl ID` <chr>,\u001b[39m\n",
      "\u001b[90m#   Pvalue <dbl>, LogFC <dbl>, Dataset <chr>, Tissue <chr>, origem <chr>\u001b[39m\n",
      "[1] \"Total number of observations:\"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "341923"
      ],
      "text/latex": [
       "341923"
      ],
      "text/markdown": [
       "341923"
      ],
      "text/plain": [
       "[1] 341923"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "File saved to: STRs_COVID_Grouped.csv\n"
     ]
    }
   ],
   "source": [
    "# Load and Clean Group Data\n",
    "df_groups <- read_csv(\"grupos.csv\") %>%\n",
    "  # Remove the \".bam\" extension from the sample column to match sample_id\n",
    "  mutate(sample_clean = str_remove(sample, \"\\\\.bam$\")) %>%\n",
    "  select(sample_clean, group) # Keep only the clean ID and the group\n",
    "\n",
    "# DEBUG: Check ID Match before joining\n",
    "cat(\"\\n--- DEBUG: Checking ID Match ---\\n\")\n",
    "ids_filtered <- unique(df_filtered$sample_id)\n",
    "ids_groups <- unique(df_groups$sample_clean)\n",
    "\n",
    "# How many IDs from df_filtered are present in the groups file?\n",
    "ids_match <- intersect(ids_filtered, ids_groups)\n",
    "n_missing <- length(setdiff(ids_filtered, ids_groups))\n",
    "\n",
    "cat(sprintf(\"Total unique samples in STR data: %d\\n\", length(ids_filtered)))\n",
    "cat(sprintf(\"Total samples with group info found: %d\\n\", length(ids_match)))\n",
    "\n",
    "if (n_missing > 0) {\n",
    "  cat(sprintf(\"WARNING: %d samples in STR data have NO group info!\\n\", n_missing))\n",
    "  # Optional: Check which ones are missing\n",
    "} else {\n",
    "  cat(\"SUCCESS: All samples have group info.\\n\")\n",
    "}\n",
    "\n",
    "# 4. Integrate (Left Join to keep STR data even if group info is missing)\n",
    "df_filtered_grouped <- df_filtered %>%\n",
    "  left_join(df_groups, by = c(\"sample_id\" = \"sample_clean\"))\n",
    "\n",
    "# 5. Preview Result\n",
    "cat(\"\\n--- Preview with Groups ---\\n\")\n",
    "df_filtered_grouped %>%\n",
    "  select(sample_id, group, everything()) %>% # Move group to the front for easy viewing\n",
    "  head() %>%\n",
    "  print()\n",
    "\n",
    "# 6. Save enriched CSV\n",
    "output_path <- \"STRs_COVID_Grouped.csv\"\n",
    "write_csv(df_filtered_grouped, output_path)\n",
    "print(\"Total number of observations:\")\n",
    "nrow(df_filtered_grouped)\n",
    "cat(sprintf(\"\\nFile saved to: %s\\n\", output_path))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b7171c2b",
   "metadata": {},
   "source": [
    "# Statistical analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c0d7e36a",
   "metadata": {},
   "source": [
    "1. Calculate Descriptive Statistics by GROUP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "5548b602",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Raw statistics saved to: ./results/allele_stats_by_group.csv\n",
      "Publication-ready table saved to: ./results/table_summary_desc.html\n"
     ]
    }
   ],
   "source": [
    "# Ensure the results directory exists\n",
    "output_dir <- \"./results\"\n",
    "if (!dir_exists(output_dir)) dir_create(output_dir)\n",
    "\n",
    "# 1. Descriptive analysis by group\n",
    "allele_stats_group <- df_filtered_grouped %>%\n",
    "  group_by(group) %>%\n",
    "  summarise(\n",
    "    n_obs = n(),\n",
    "    across(c(allele1_est, allele2_est), \n",
    "           list(mean = ~mean(.x, na.rm = TRUE), \n",
    "                median = ~median(.x, na.rm = TRUE), \n",
    "                sd = ~sd(.x, na.rm = TRUE),\n",
    "                iqr = ~IQR(.x, na.rm = TRUE)),\n",
    "           .names = \"{.col}_{.fn}\")\n",
    "  ) %>%\n",
    "  pivot_longer(cols = -group, names_to = \"Metric\", values_to = \"value\")\n",
    "\n",
    "# Save raw CSV data first\n",
    "stats_csv_path <- file.path(output_dir, \"allele_stats_by_group.csv\")\n",
    "write_csv(allele_stats_group, stats_csv_path)\n",
    "\n",
    "cat(sprintf(\"Raw statistics saved to: %s\\n\", stats_csv_path))\n",
    "\n",
    "# ==============================================================================\n",
    "# 2. Format Table for Publication (GT Package)\n",
    "# ==============================================================================\n",
    "\n",
    "# Reshape to Wide Format for display (Metrics in rows, Groups in columns)\n",
    "df_wide_pub <- allele_stats_group %>%\n",
    "  pivot_wider(\n",
    "    names_from = group,\n",
    "    values_from = value\n",
    "  ) %>%\n",
    "  # Rename columns to match the target style\n",
    "  rename(\n",
    "    'Case Group' = case,\n",
    "    'Control Group' = control\n",
    "  )\n",
    "\n",
    "# Translate/Format Metric names explicitly\n",
    "# Using proper English scientific notation\n",
    "df_wide_pub <- df_wide_pub %>%\n",
    "  mutate(Metric = case_when(\n",
    "    Metric == \"allele1_est_mean\"   ~ \"Mean (Allele 1)\",\n",
    "    Metric == \"allele1_est_median\" ~ \"Median (Allele 1)\",\n",
    "    Metric == \"allele1_est_sd\"     ~ \"SD (Allele 1)\",\n",
    "    Metric == \"allele1_est_iqr\"    ~ \"IQR (Allele 1)\",\n",
    "    \n",
    "    Metric == \"allele2_est_mean\"   ~ \"Mean (Allele 2)\",\n",
    "    Metric == \"allele2_est_median\" ~ \"Median (Allele 2)\",\n",
    "    Metric == \"allele2_est_sd\"     ~ \"SD (Allele 2)\",\n",
    "    Metric == \"allele2_est_iqr\"    ~ \"IQR (Allele 2)\",\n",
    "    \n",
    "    Metric == \"n_obs\"              ~ \"Total Observations\",\n",
    "    TRUE ~ Metric # Keep others as is if any\n",
    "  )) %>%\n",
    "  # Custom Sort order (Optional: Obs -> Allele 1 -> Allele 2)\n",
    "  arrange(factor(Metric, levels = c(\n",
    "    \"Total Observations\",\n",
    "    \"Mean (Allele 1)\", \"Median (Allele 1)\", \"SD (Allele 1)\", \"IQR (Allele 1)\",\n",
    "    \"Mean (Allele 2)\", \"Median (Allele 2)\", \"SD (Allele 2)\", \"IQR (Allele 2)\"\n",
    "  )))\n",
    "\n",
    "# Create GT Table\n",
    "tabela_gt <- gt(df_wide_pub) %>%\n",
    "  # Number formatting\n",
    "  fmt_number(\n",
    "    columns = -Metric,      \n",
    "    decimals = 2,\n",
    "    use_seps = TRUE           \n",
    "  ) %>%\n",
    "  # Special formatting for Observation count (no decimals)\n",
    "  fmt_number(\n",
    "    columns = -Metric,\n",
    "    rows = Metric == \"Total Observations\",\n",
    "    decimals = 0\n",
    "  ) %>%\n",
    "  # Table Header\n",
    "  tab_header(\n",
    "    title = \"Descriptive Statistics of STR Metrics\",\n",
    "    subtitle = \"Comparison between Case and Control Groups\"\n",
    "  ) %>%\n",
    "  # Column Labels\n",
    "  cols_label(\n",
    "    Metric = \"Metric\",\n",
    "    `Case Group` = \"Case Group\",\n",
    "    `Control Group` = \"Control Group\"\n",
    "  ) %>%\n",
    "  # Footnotes\n",
    "  tab_footnote(\n",
    "    footnote = \"SD: Standard Deviation; IQR: Interquartile Range.\",\n",
    "    locations = cells_column_labels(columns = Metric)\n",
    "  ) %>%\n",
    "  # Style: Bold Metric names\n",
    "  tab_style(\n",
    "    style = cell_text(weight = \"bold\"),\n",
    "    locations = cells_body(columns = Metric)\n",
    "  )\n",
    "\n",
    "# 3. Save as HTML\n",
    "html_path <- file.path(output_dir, \"table_summary_desc.html\")\n",
    "gtsave(tabela_gt, html_path)\n",
    "\n",
    "cat(sprintf(\"Publication-ready table saved to: %s\\n\", html_path))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e3980278",
   "metadata": {},
   "source": [
    "2. Calculate data distribution between loci and groups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "7f962f3b",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "--- Normality Test Summary (Counts of Loci) ---\n",
      "\u001b[90m# A tibble: 6 × 5\u001b[39m\n",
      "  variable     group   `Insufficient Data/Constant` `Non-Normal` Normal\n",
      "  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m        \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m                          \u001b[3m\u001b[90m<int>\u001b[39m\u001b[23m        \u001b[3m\u001b[90m<int>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<int>\u001b[39m\u001b[23m\n",
      "\u001b[90m1\u001b[39m allele1_est  case                             530         \u001b[4m1\u001b[24m364   \u001b[4m1\u001b[24m583\n",
      "\u001b[90m2\u001b[39m allele1_est  control                          242         \u001b[4m2\u001b[24m332   \u001b[4m1\u001b[24m396\n",
      "\u001b[90m3\u001b[39m allele2_est  case                             530         \u001b[4m1\u001b[24m588   \u001b[4m1\u001b[24m359\n",
      "\u001b[90m4\u001b[39m allele2_est  control                          242         \u001b[4m2\u001b[24m961    767\n",
      "\u001b[90m5\u001b[39m mean_alleles case                             530         \u001b[4m1\u001b[24m545   \u001b[4m1\u001b[24m402\n",
      "\u001b[90m6\u001b[39m mean_alleles control                          242         \u001b[4m2\u001b[24m744    984\n",
      "\n",
      "Detailed results saved to: ./results/shapiro_normality_per_locus_allele_group.csv\n"
     ]
    }
   ],
   "source": [
    "# 0. Data Preparation\n",
    "# Calculate mean_alleles\n",
    "df_normality <- df_filtered_grouped %>%\n",
    "  mutate(mean_alleles = (allele1_est + allele2_est) / 2)\n",
    "\n",
    "# 1. Define Function to Run Shapiro Test Safely\n",
    "safe_shapiro <- function(x) {\n",
    "  # Remove NAs\n",
    "  clean_x <- x[!is.na(x)]\n",
    "  n <- length(clean_x)\n",
    "  \n",
    "  # Shapiro requires 3 to 5000 samples\n",
    "  if (n < 3) {\n",
    "    return(tibble(statistic = NA, p.value = NA, n_obs = n, status = \"Too few samples\"))\n",
    "  }\n",
    "  if (n > 5000) {\n",
    "    # If a single locus has >5000 samples, we subsample to allow the test to run\n",
    "    set.seed(123)\n",
    "    clean_x <- sample(clean_x, 5000)\n",
    "  }\n",
    "  \n",
    "  # Constant variance check (all values identical)\n",
    "  if (var(clean_x) == 0) {\n",
    "    return(tibble(statistic = NA, p.value = 1, n_obs = n, status = \"Constant Variance\"))\n",
    "  }\n",
    "  \n",
    "  tryCatch({\n",
    "    res <- shapiro_test(clean_x)\n",
    "    return(tibble(statistic = res$statistic, p.value = res$p.value, n_obs = n, status = \"OK\"))\n",
    "  }, error = function(e) return(tibble(statistic = NA, p.value = NA, n_obs = n, status = \"Error\")))\n",
    "}\n",
    "\n",
    "# 2. Run Normality Tests (Per Locus, Per Group, Per Allele/Variable)\n",
    "normality_results <- df_normality %>%\n",
    "  select(STRs_ID, group, allele1_est, allele2_est, mean_alleles) %>%\n",
    "  pivot_longer(cols = c(allele1_est, allele2_est, mean_alleles), \n",
    "               names_to = \"variable\", \n",
    "               values_to = \"value\") %>%\n",
    "  # VITAL: Grouping by STRs_ID prevents mixing different loci distributions\n",
    "  group_by(STRs_ID, group, variable) %>%\n",
    "  nest() %>%\n",
    "  mutate(shapiro = map(data, ~safe_shapiro(.x$value))) %>%\n",
    "  unnest(shapiro) %>%\n",
    "  ungroup() %>%\n",
    "  mutate(\n",
    "    normality = case_when(\n",
    "      is.na(p.value) ~ \"Insufficient Data/Constant\",\n",
    "      p.value > 0.05 ~ \"Normal\",\n",
    "      p.value <= 0.05 ~ \"Non-Normal\" # Significant deviation from normal curve\n",
    "    )\n",
    "  )\n",
    "\n",
    "# 3. Summary of Results\n",
    "cat(\"\\n--- Normality Test Summary (Counts of Loci) ---\\n\")\n",
    "\n",
    "# Count how many loci are Normal vs Non-Normal for each Variable/Group combination\n",
    "summary_table <- normality_results %>%\n",
    "  count(variable, group, normality) %>%\n",
    "  pivot_wider(names_from = normality, values_from = n, values_fill = 0)\n",
    "\n",
    "print(summary_table)\n",
    "\n",
    "# 4. Save Results\n",
    "output_dir <- \"./results\"\n",
    "if (!dir.exists(output_dir)) dir.create(output_dir)\n",
    "\n",
    "# Saving the detailed per-locus table\n",
    "write_csv(normality_results, file.path(output_dir, \"shapiro_normality_per_locus_allele_group.csv\"))\n",
    "\n",
    "cat(sprintf(\"\\nDetailed results saved to: %s/shapiro_normality_per_locus_allele_group.csv\\n\", output_dir))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "47811dac",
   "metadata": {},
   "source": [
    "3. Mann-whitney analysis between the major allele length"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "3898db78",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--- Starting Data Preparation (Target: Allele 2) ---\n",
      "Data prepared. Deduplicated rows: 144796 \n",
      "Using metric: Allele 2\n",
      "Eligible loci for testing: 1922\n",
      "\n",
      "--- Creating Metadata Dictionary ---\n",
      "\n",
      "--- Running Statistical Tests ---\n",
      "\n",
      "--- Consolidating Results ---\n",
      "\n",
      "=== FINAL SUMMARY (Allele 2) ===\n",
      "Loci analyzed: 1922 \n",
      "Significant (p.adj < 0.05): 0 \n",
      "\u001b[90m# A tibble: 6 × 19\u001b[39m\n",
      "  locus_definition  STRs_ID    gene_name chrom  start region .y.   group1 group2\n",
      "  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m             \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m      \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m     \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \n",
      "\u001b[90m1\u001b[39m chr1:8018387:C    chr1:8018… ERRFI1    1     8.02\u001b[90me\u001b[39m6 intron targ… case   contr…\n",
      "\u001b[90m2\u001b[39m chr13:91965100:C  chr13:919… GPC5      13    9.20\u001b[90me\u001b[39m7 intron targ… case   contr…\n",
      "\u001b[90m3\u001b[39m chr15:38983802:C  chr15:389… AC013652… 15    3.90\u001b[90me\u001b[39m7 intron targ… case   contr…\n",
      "\u001b[90m4\u001b[39m chr6:156000320:C  chr6:1560… AL589693… 6     1.56\u001b[90me\u001b[39m8 intron targ… case   contr…\n",
      "\u001b[90m5\u001b[39m chr3:169212653:AT chr3:1692… MECOM     3     1.69\u001b[90me\u001b[39m8 intron targ… case   contr…\n",
      "\u001b[90m6\u001b[39m chr17:57467841:G  chr17:574… MSI2      17    5.75\u001b[90me\u001b[39m7 intron targ… case   contr…\n",
      "\u001b[90m# ℹ 10 more variables: n1 <int>, n2 <int>, statistic <dbl>, p <dbl>,\u001b[39m\n",
      "\u001b[90m#   median_val_case <dbl>, median_val_control <dbl>, iqr_val_case <dbl>,\u001b[39m\n",
      "\u001b[90m#   iqr_val_control <dbl>, p.adj <dbl>, p.adj.signif <chr>\u001b[39m\n"
     ]
    }
   ],
   "source": [
    "# =======================================================\n",
    "# 0. DATA PREPARATION & DEDUPLICATION (STEP ZERO)\n",
    "# =======================================================\n",
    "cat(\"--- Starting Data Preparation (Target: Allele 2) ---\\n\")\n",
    "\n",
    "# 1. Create Identifiers and Select Metric\n",
    "df_prepared <- df_filtered_grouped %>%\n",
    "  # Create the Locus ID\n",
    "  mutate(locus_definition = str_remove(STRs_ID, \":[^:]+$\")) %>%\n",
    "  \n",
    "  # DEFINING THE METRIC OF INTEREST: ALLELE 2\n",
    "  mutate(target_metric = allele2_est)\n",
    "\n",
    "# 2. Deduplication (Ensure Uniqueness: 1 row per Sample per Locus)\n",
    "df_unique_samples <- df_prepared %>%\n",
    "  group_by(locus_definition, sample_id, group) %>%\n",
    "  summarise(\n",
    "    # Mean of the target metric (in case of technical replicates)\n",
    "    target_metric = mean(target_metric, na.rm = TRUE),\n",
    "    \n",
    "    # Preserve metadata\n",
    "    gene_name = first(gene_name),\n",
    "    chrom     = first(chrom),\n",
    "    start     = first(start),\n",
    "    region    = first(region),   \n",
    "    STRs_ID   = first(STRs_ID),  \n",
    "    \n",
    "    .groups = \"drop\"\n",
    "  )\n",
    "\n",
    "# Update main dataframe\n",
    "df_per_locus <- df_unique_samples\n",
    "\n",
    "cat(\"Data prepared. Deduplicated rows:\", nrow(df_per_locus), \"\\n\")\n",
    "cat(\"Using metric: Allele 2\\n\")\n",
    "\n",
    "\n",
    "# =======================================================\n",
    "# 1. FILTER ELIGIBLE LOCI\n",
    "# =======================================================\n",
    "loci_counts <- df_per_locus %>%\n",
    "  group_by(locus_definition, group) %>%\n",
    "  summarise(n = sum(!is.na(target_metric)), .groups = \"drop\") %>%\n",
    "  pivot_wider(names_from = group, values_from = n, values_fill = 0)\n",
    "\n",
    "# Filter loci with enough samples\n",
    "loci_to_test <- loci_counts %>%\n",
    "  filter(case >= 5 & control >= 5) %>%\n",
    "  pull(locus_definition)\n",
    "\n",
    "cat(sprintf(\"Eligible loci for testing: %d\\n\", length(loci_to_test)))\n",
    "\n",
    "\n",
    "# =======================================================\n",
    "# 2. METADATA LOOKUP TABLE\n",
    "# =======================================================\n",
    "cat(\"\\n--- Creating Metadata Dictionary ---\\n\")\n",
    "\n",
    "loci_metadata <- df_per_locus %>%\n",
    "  # <--- ADICIONADO: STRs_ID no select\n",
    "  select(locus_definition, gene_name, chrom, start, region, STRs_ID) %>%\n",
    "  group_by(locus_definition) %>%\n",
    "  summarise(\n",
    "    gene_name = first(na.omit(gene_name)),\n",
    "    chrom     = first(na.omit(chrom)),\n",
    "    start     = first(na.omit(start)),\n",
    "    region    = first(na.omit(region)),\n",
    "    STRs_ID   = first(na.omit(STRs_ID)), \n",
    "    n_versions = n_distinct(gene_name, na.rm = TRUE),\n",
    "    .groups = \"drop\"\n",
    "  )\n",
    "\n",
    "if (any(loci_metadata$n_versions > 1)) {\n",
    "  warning(\"Some loci have multiple gene names. Using the first valid occurrence.\")\n",
    "}\n",
    "loci_metadata <- loci_metadata %>% select(-n_versions)\n",
    "\n",
    "\n",
    "# =======================================================\n",
    "# 3. SAFE STATISTICAL FUNCTION\n",
    "# =======================================================\n",
    "safe_wilcox_metric <- function(data) {\n",
    "  tryCatch({\n",
    "    # Check for variance in target_metric\n",
    "    if (n_distinct(data$target_metric) < 2) return(NULL)\n",
    "    \n",
    "    # Mann-Whitney test on target_metric\n",
    "    test_res <- data %>% wilcox_test(target_metric ~ group)\n",
    "    \n",
    "    # Stats Summary\n",
    "    stats_summary <- data %>%\n",
    "      group_by(group) %>%\n",
    "      summarise(\n",
    "        median_val = median(target_metric, na.rm = TRUE),\n",
    "        iqr_val    = IQR(target_metric, na.rm = TRUE),\n",
    "        .groups = \"drop\"\n",
    "      ) %>%\n",
    "      pivot_wider(\n",
    "        names_from = group,\n",
    "        values_from = c(median_val, iqr_val),\n",
    "        names_glue = \"{.value}_{group}\"\n",
    "      )\n",
    "    \n",
    "    return(bind_cols(test_res, stats_summary))\n",
    "  }, error = function(e) { return(NULL) })\n",
    "}\n",
    "\n",
    "\n",
    "# =======================================================\n",
    "# 4. RUN TESTS\n",
    "# =======================================================\n",
    "cat(\"\\n--- Running Statistical Tests ---\\n\")\n",
    "\n",
    "nested_results <- df_per_locus %>%\n",
    "  filter(locus_definition %in% loci_to_test) %>%\n",
    "  \n",
    "  # Select metric of interest\n",
    "  select(locus_definition, sample_id, group, target_metric) %>%\n",
    "  \n",
    "  group_by(locus_definition) %>%\n",
    "  nest() %>%\n",
    "  \n",
    "  mutate(model = map(data, safe_wilcox_metric)) %>%\n",
    "  \n",
    "  filter(!map_lgl(model, is.null))\n",
    "\n",
    "\n",
    "# =======================================================\n",
    "# 5. METADATA REINCORPORATION\n",
    "# =======================================================\n",
    "cat(\"\\n--- Consolidating Results ---\\n\")\n",
    "\n",
    "# Statistical Results\n",
    "results_mw_allele2 <- nested_results %>%\n",
    "  select(-data) %>%\n",
    "  unnest(model) %>%\n",
    "  ungroup() %>%\n",
    "  left_join(loci_metadata, by = \"locus_definition\") %>%\n",
    "  select(locus_definition, STRs_ID, gene_name, chrom, start, region, everything()) %>%\n",
    "  adjust_pvalue(method = \"BH\") %>%\n",
    "  add_significance(\"p.adj\") %>%\n",
    "  arrange(p)\n",
    "\n",
    "# Raw Observations\n",
    "raw_obs_allele2 <- nested_results %>%\n",
    "  ungroup() %>%\n",
    "  select(locus_definition, data) %>%\n",
    "  unnest(data) %>%\n",
    "  left_join(loci_metadata, by = \"locus_definition\") %>%\n",
    "  select(locus_definition, STRs_ID, gene_name, sample_id, group, target_metric)\n",
    "\n",
    "\n",
    "# =======================================================\n",
    "# 6. OUTPUT\n",
    "# =======================================================\n",
    "cat(\"\\n=== FINAL SUMMARY (Allele 2) ===\\n\")\n",
    "cat(\"Loci analyzed:\", nrow(results_mw_allele2), \"\\n\")\n",
    "cat(\"Significant (p.adj < 0.05):\", sum(results_mw_allele2$p.adj < 0.05, na.rm = TRUE), \"\\n\")\n",
    "\n",
    "print(head(results_mw_allele2))\n",
    "\n",
    "# Save with specific names\n",
    "if(!dir.exists(\"results\")) dir.create(\"results\")\n",
    "write_csv(results_mw_allele2, \"results/allele2_stats_results_mann-whitney.csv\")\n",
    "write_csv(raw_obs_allele2, \"results/allele2_raw_observations_mann-whitney.csv\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6545ca8e",
   "metadata": {},
   "source": [
    "4. Mann-whitney analysis between the mean length of alleles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "83120496",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--- Starting No-Overlap Analysis for Mean Alleles ---\n",
      "--- Preparing data with Mean Alleles metric ---\n",
      "Data prepared. Rows: 144796 \n",
      "\n",
      "--- Computing Statistics per Locus ---\n",
      "Locus stats calculated. Rows: 7447 \n",
      "\u001b[90m# A tibble: 6 × 8\u001b[39m\n",
      "  locus_definition  group   median_mean min_mean max_mean mean_value sd_mean\n",
      "  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m             \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m         \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m    \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m    \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m      \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m   \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m\n",
      "\u001b[90m1\u001b[39m chr10:10857639:AT case           8.43     5.78     12.2       8.91    2.13\n",
      "\u001b[90m2\u001b[39m chr10:10857639:AT control        9.49     2.86     19.0      10.1     3.34\n",
      "\u001b[90m3\u001b[39m chr10:10994724:C  case          20.7     14.8      31.0      21.1     5.09\n",
      "\u001b[90m4\u001b[39m chr10:10994724:C  control       17.1     13.5      30.5      18.7     3.84\n",
      "\u001b[90m5\u001b[39m chr10:11005244:AG case           8.13     5.96     14.6       8.75    2.78\n",
      "\u001b[90m6\u001b[39m chr10:11005244:AG control       10.1      2.33     18.5       9.81    3.55\n",
      "\u001b[90m# ℹ 1 more variable: n_samples <int>\u001b[39m\n",
      "\n",
      "--- Identifying No-Overlap Variants ---\n",
      "Total loci: 4046 \n",
      "Loci with NO-OVERLAP (min 3 samples per group): 2 \n",
      "\n",
      "--- Creating Statistics CSV ---\n",
      "\u001b[90m# A tibble: 10 × 20\u001b[39m\n",
      "   locus_definition  STRs_ID      gene_name region chrom  start median_mean_case\n",
      "   \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m             \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m        \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m     \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m            \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m\n",
      "\u001b[90m 1\u001b[39m chr12:77471132:A  chr12:77471… NAV3      intron 12    7.75\u001b[90me\u001b[39m7            14.9 \n",
      "\u001b[90m 2\u001b[39m chr3:71057291:T   chr3:710572… FOXP1     intron 3     7.11\u001b[90me\u001b[39m7            14.3 \n",
      "\u001b[90m 3\u001b[39m chr10:10857639:AT chr10:10857… CELF2     intron 10    1.09\u001b[90me\u001b[39m7             8.43\n",
      "\u001b[90m 4\u001b[39m chr10:10994724:C  chr10:10994… CELF2     intron 10    1.10\u001b[90me\u001b[39m7            20.7 \n",
      "\u001b[90m 5\u001b[39m chr10:11005244:AG chr10:11005… CELF2     intron 10    1.10\u001b[90me\u001b[39m7             8.13\n",
      "\u001b[90m 6\u001b[39m chr10:11032146:A  chr10:11032… CELF2     intron 10    1.10\u001b[90me\u001b[39m7            14.4 \n",
      "\u001b[90m 7\u001b[39m chr10:11086578:A  chr10:11086… CELF2     intron 10    1.11\u001b[90me\u001b[39m7            \u001b[31mNA\u001b[39m   \n",
      "\u001b[90m 8\u001b[39m chr10:11103487:T  chr10:11103… CELF2     intron 10    1.11\u001b[90me\u001b[39m7            23.5 \n",
      "\u001b[90m 9\u001b[39m chr10:11119864:G  chr10:11119… CELF2     intron 10    1.11\u001b[90me\u001b[39m7            30.7 \n",
      "\u001b[90m10\u001b[39m chr10:11165056:C  chr10:11165… CELF2     five_… 10    1.12\u001b[90me\u001b[39m7            30.9 \n",
      "\u001b[90m# ℹ 13 more variables: min_mean_case <dbl>, max_mean_case <dbl>,\u001b[39m\n",
      "\u001b[90m#   mean_value_case <dbl>, sd_mean_case <dbl>, n_samples_case <int>,\u001b[39m\n",
      "\u001b[90m#   median_mean_control <dbl>, min_mean_control <dbl>, max_mean_control <dbl>,\u001b[39m\n",
      "\u001b[90m#   mean_value_control <dbl>, sd_mean_control <dbl>, n_samples_control <int>,\u001b[39m\n",
      "\u001b[90m#   has_no_overlap <lgl>, overlap_type <chr>\u001b[39m\n",
      "✓ Saved: mean_alleles_no_overlap_statistics_per_locus.csv\n",
      "\n",
      "--- Creating No-Overlap Variants CSV ---\n",
      "\u001b[90m# A tibble: 2 × 15\u001b[39m\n",
      "  locus_definition STRs_ID        gene_name region chrom  start median_mean_case\n",
      "  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m            \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m          \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m     \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m            \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m\n",
      "\u001b[90m1\u001b[39m chr12:77471132:A chr12:7747113… NAV3      intron 12    7.75\u001b[90me\u001b[39m7             14.9\n",
      "\u001b[90m2\u001b[39m chr3:71057291:T  chr3:71057291… FOXP1     intron 3     7.11\u001b[90me\u001b[39m7             14.3\n",
      "\u001b[90m# ℹ 8 more variables: min_mean_case <dbl>, max_mean_case <dbl>,\u001b[39m\n",
      "\u001b[90m#   n_samples_case <int>, median_mean_control <dbl>, min_mean_control <dbl>,\u001b[39m\n",
      "\u001b[90m#   max_mean_control <dbl>, n_samples_control <int>, overlap_type <chr>\u001b[39m\n",
      "✓ Saved: mean_alleles_no_overlap_variants_only.csv\n",
      "\n",
      "=== NO-OVERLAP ANALYSIS SUMMARY (Mean Alleles) ===\n",
      "Total loci analyzed: 4046 \n",
      "Loci with NO-OVERLAP (min 3 samples per group): 2 \n",
      "\n",
      "No-overlap breakdown:\n",
      "\u001b[90m# A tibble: 1 × 2\u001b[39m\n",
      "  overlap_type   count\n",
      "  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m          \u001b[3m\u001b[90m<int>\u001b[39m\u001b[23m\n",
      "\u001b[90m1\u001b[39m control_higher     2\n"
     ]
    }
   ],
   "source": [
    "# =======================================================\n",
    "# ANALYSIS: MEAN ALLELES NO-OVERLAP DETECTION (CASE vs CONTROL)\n",
    "# =======================================================\n",
    "cat(\"--- Starting No-Overlap Analysis for Mean Alleles ---\\n\")\n",
    "\n",
    "\n",
    "# =======================================================\n",
    "# 1. PREPARE DATA FOR NO-OVERLAP ANALYSIS\n",
    "# =======================================================\n",
    "\n",
    "# Force recreation with the correct metric (Mean Alleles)\n",
    "cat(\"--- Preparing data with Mean Alleles metric ---\\n\")\n",
    "\n",
    "df_prepared <- df_filtered_grouped %>%\n",
    "  mutate(locus_definition = str_remove(STRs_ID, \":[^:]+$\")) %>%\n",
    "  mutate(mean_alleles_val = (allele1_est + allele2_est) / 2)\n",
    "\n",
    "df_per_locus <- df_prepared %>%\n",
    "  group_by(locus_definition, sample_id, group) %>%\n",
    "  summarise(\n",
    "    mean_alleles_val = mean(mean_alleles_val, na.rm = TRUE),\n",
    "    gene_name = first(gene_name),\n",
    "    chrom     = first(chrom),\n",
    "    start     = first(start),\n",
    "    region    = first(region),\n",
    "    STRs_ID   = first(STRs_ID),\n",
    "    .groups = \"drop\"\n",
    "  )\n",
    "\n",
    "cat(\"Data prepared. Rows:\", nrow(df_per_locus), \"\\n\")\n",
    "\n",
    "\n",
    "# =======================================================\n",
    "# 2. CALCULATE STATISTICS PER LOCUS AND GROUP\n",
    "# =======================================================\n",
    "cat(\"\\n--- Computing Statistics per Locus ---\\n\")\n",
    "\n",
    "locus_stats <- df_per_locus %>%\n",
    "  filter(!is.na(mean_alleles_val)) %>%\n",
    "  group_by(locus_definition, group) %>%\n",
    "  summarise(\n",
    "    median_mean = median(mean_alleles_val, na.rm = TRUE),\n",
    "    min_mean    = min(mean_alleles_val, na.rm = TRUE),\n",
    "    max_mean    = max(mean_alleles_val, na.rm = TRUE),\n",
    "    mean_value  = mean(mean_alleles_val, na.rm = TRUE),\n",
    "    sd_mean     = sd(mean_alleles_val, na.rm = TRUE),\n",
    "    n_samples   = n(),\n",
    "    .groups = \"drop\"\n",
    "  )\n",
    "\n",
    "# Display the structure\n",
    "cat(\"Locus stats calculated. Rows:\", nrow(locus_stats), \"\\n\")\n",
    "print(head(locus_stats))\n",
    "\n",
    "\n",
    "# =======================================================\n",
    "# 3. IDENTIFY NO-OVERLAP VARIANTS\n",
    "# =======================================================\n",
    "cat(\"\\n--- Identifying No-Overlap Variants ---\\n\")\n",
    "\n",
    "# Pivot to compare case vs control side by side\n",
    "locus_comparison <- locus_stats %>%\n",
    "  pivot_wider(\n",
    "    id_cols = locus_definition,\n",
    "    names_from = group,\n",
    "    values_from = c(median_mean, min_mean, max_mean, mean_value, sd_mean, n_samples),\n",
    "    names_glue = \"{.value}_{group}\"\n",
    "  )\n",
    "\n",
    "# Detect no-overlap WITH MINIMUM SAMPLE SIZE REQUIREMENT\n",
    "# Only consider variants with at least N samples in BOTH groups\n",
    "MIN_SAMPLES_PER_GROUP <- 3  # Adjust this threshold as needed\n",
    "\n",
    "locus_comparison <- locus_comparison %>%\n",
    "  mutate(\n",
    "    has_no_overlap = (n_samples_case >= MIN_SAMPLES_PER_GROUP & \n",
    "                      n_samples_control >= MIN_SAMPLES_PER_GROUP) &\n",
    "                     ((min_mean_case > max_mean_control) | \n",
    "                      (min_mean_control > max_mean_case)),\n",
    "    \n",
    "    overlap_type = case_when(\n",
    "      n_samples_case < MIN_SAMPLES_PER_GROUP ~ \"insufficient_case_samples\",\n",
    "      n_samples_control < MIN_SAMPLES_PER_GROUP ~ \"insufficient_control_samples\",\n",
    "      min_mean_case > max_mean_control ~ \"case_higher\",\n",
    "      min_mean_control > max_mean_case ~ \"control_higher\",\n",
    "      TRUE ~ \"overlap\"\n",
    "    )\n",
    "  )\n",
    "\n",
    "# Recover metadata\n",
    "locus_with_metadata <- locus_comparison %>%\n",
    "  left_join(\n",
    "    df_per_locus %>% \n",
    "      select(locus_definition, gene_name, region, STRs_ID, chrom, start) %>%\n",
    "      distinct(),\n",
    "    by = \"locus_definition\"\n",
    "  )\n",
    "\n",
    "cat(\"Total loci:\", nrow(locus_with_metadata), \"\\n\")\n",
    "cat(\"Loci with NO-OVERLAP (min\", MIN_SAMPLES_PER_GROUP, \"samples per group):\", \n",
    "    sum(locus_with_metadata$has_no_overlap, na.rm = TRUE), \"\\n\")\n",
    "cat(\"\\n\")\n",
    "\n",
    "\n",
    "# =======================================================\n",
    "# 4. CREATE OUTPUT CSV 1: DETAILED STATISTICS PER LOCUS\n",
    "# =======================================================\n",
    "cat(\"--- Creating Statistics CSV ---\\n\")\n",
    "\n",
    "csv_statistics <- locus_with_metadata %>%\n",
    "  select(\n",
    "    locus_definition,\n",
    "    STRs_ID,\n",
    "    gene_name,\n",
    "    region,\n",
    "    chrom,\n",
    "    start,\n",
    "    \n",
    "    # Case group\n",
    "    median_mean_case,\n",
    "    min_mean_case,\n",
    "    max_mean_case,\n",
    "    mean_value_case,\n",
    "    sd_mean_case,\n",
    "    n_samples_case,\n",
    "    \n",
    "    # Control group\n",
    "    median_mean_control,\n",
    "    min_mean_control,\n",
    "    max_mean_control,\n",
    "    mean_value_control,\n",
    "    sd_mean_control,\n",
    "    n_samples_control,\n",
    "    \n",
    "    # Overlap info\n",
    "    has_no_overlap,\n",
    "    overlap_type\n",
    "  ) %>%\n",
    "  arrange(desc(has_no_overlap), locus_definition)\n",
    "\n",
    "print(head(csv_statistics, 10))\n",
    "\n",
    "# Create results directory if it doesn't exist\n",
    "if(!dir.exists(\"results\")) dir.create(\"results\")\n",
    "\n",
    "write_csv(\n",
    "  csv_statistics,\n",
    "  \"results/mean_alleles_no_overlap_statistics_per_locus.csv\"\n",
    ")\n",
    "\n",
    "cat(\"✓ Saved: mean_alleles_no_overlap_statistics_per_locus.csv\\n\")\n",
    "\n",
    "\n",
    "# =======================================================\n",
    "# 5. CREATE OUTPUT CSV 2: ONLY NO-OVERLAP VARIANTS\n",
    "# =======================================================\n",
    "cat(\"\\n--- Creating No-Overlap Variants CSV ---\\n\")\n",
    "\n",
    "csv_no_overlap_only <- locus_with_metadata %>%\n",
    "  filter(has_no_overlap == TRUE) %>%\n",
    "  select(\n",
    "    locus_definition,\n",
    "    STRs_ID,\n",
    "    gene_name,\n",
    "    region,\n",
    "    chrom,\n",
    "    start,\n",
    "    \n",
    "    # Case group\n",
    "    median_mean_case,\n",
    "    min_mean_case,\n",
    "    max_mean_case,\n",
    "    n_samples_case,\n",
    "    \n",
    "    # Control group\n",
    "    median_mean_control,\n",
    "    min_mean_control,\n",
    "    max_mean_control,\n",
    "    n_samples_control,\n",
    "    \n",
    "    # Interpretation\n",
    "    overlap_type\n",
    "  ) %>%\n",
    "  arrange(overlap_type, locus_definition)\n",
    "\n",
    "print(csv_no_overlap_only)\n",
    "\n",
    "write_csv(\n",
    "  csv_no_overlap_only,\n",
    "  \"results/mean_alleles_no_overlap_variants_only.csv\"\n",
    ")\n",
    "\n",
    "cat(\"✓ Saved: mean_alleles_no_overlap_variants_only.csv\\n\")\n",
    "\n",
    "\n",
    "# =======================================================\n",
    "# 6. SUMMARY REPORT\n",
    "# =======================================================\n",
    "cat(\"\\n=== NO-OVERLAP ANALYSIS SUMMARY (Mean Alleles) ===\\n\")\n",
    "cat(\"Total loci analyzed:\", nrow(csv_statistics), \"\\n\")\n",
    "cat(\"Loci with NO-OVERLAP (min\", MIN_SAMPLES_PER_GROUP, \"samples per group):\", \n",
    "    nrow(csv_no_overlap_only), \"\\n\")\n",
    "cat(\"\\nNo-overlap breakdown:\\n\")\n",
    "\n",
    "if(nrow(csv_no_overlap_only) > 0) {\n",
    "  breakdown <- csv_no_overlap_only %>%\n",
    "    group_by(overlap_type) %>%\n",
    "    summarise(\n",
    "      count = n(),\n",
    "      .groups = \"drop\"\n",
    "    )\n",
    "  \n",
    "  print(breakdown)\n",
    "} else {\n",
    "  cat(\"No variants without overlap found.\\n\")\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "33158a50",
   "metadata": {},
   "source": [
    "5. Identification of loci without overlap in relation of the major allele"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "675ed9c1",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--- Starting No-Overlap Analysis for Allele 2 ---\n",
      "--- Preparing data with Allele 2 metric ---\n",
      "Data prepared. Rows: 144796 \n",
      "\n",
      "--- Computing Statistics per Locus ---\n",
      "Locus stats calculated. Rows: 7447 \n",
      "\u001b[90m# A tibble: 6 × 8\u001b[39m\n",
      "  locus_definition  group   median_allele2 min_allele2 max_allele2 mean_allele2\n",
      "  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m             \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m            \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m       \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m       \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m        \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m\n",
      "\u001b[90m1\u001b[39m chr10:10857639:AT case              16.8        14          21.4         17.8\n",
      "\u001b[90m2\u001b[39m chr10:10857639:AT control           17.7        10.7        41.4         19.3\n",
      "\u001b[90m3\u001b[39m chr10:10994724:C  case              41.4        29.6        62.0         42.3\n",
      "\u001b[90m4\u001b[39m chr10:10994724:C  control           34.2        27.0        61.0         37.4\n",
      "\u001b[90m5\u001b[39m chr10:11005244:AG case              16.8        13.4        29.2         18.0\n",
      "\u001b[90m6\u001b[39m chr10:11005244:AG control           19.2        10.5        37.0         20.2\n",
      "\u001b[90m# ℹ 2 more variables: sd_allele2 <dbl>, n_samples <int>\u001b[39m\n",
      "\n",
      "--- Identifying No-Overlap Variants ---\n",
      "Total loci: 4046 \n",
      "Loci with NO-OVERLAP (min 3 samples per group): 3 \n",
      "\n",
      "Breakdown by category:\n",
      "  case_higher: 1 \n",
      "  control_higher: 2 \n",
      "  insufficient_case_samples: 914 \n",
      "  insufficient_control_samples: 252 \n",
      "  overlap: 2877 \n",
      "\n",
      "--- Creating Statistics CSV ---\n",
      "\u001b[90m# A tibble: 10 × 20\u001b[39m\n",
      "   locus_definition  STRs_ID   gene_name region chrom  start median_allele2_case\n",
      "   \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m             \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m     \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m     \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m               \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m\n",
      "\u001b[90m 1\u001b[39m chr12:77471132:A  chr12:77… NAV3      intron 12    7.75\u001b[90me\u001b[39m7                31.7\n",
      "\u001b[90m 2\u001b[39m chr3:41935209:T   chr3:419… ULK4      intron 3     4.19\u001b[90me\u001b[39m7                54.0\n",
      "\u001b[90m 3\u001b[39m chr3:71057291:T   chr3:710… FOXP1     intron 3     7.11\u001b[90me\u001b[39m7                25.8\n",
      "\u001b[90m 4\u001b[39m chr10:10857639:AT chr10:10… CELF2     intron 10    1.09\u001b[90me\u001b[39m7                16.8\n",
      "\u001b[90m 5\u001b[39m chr10:10994724:C  chr10:10… CELF2     intron 10    1.10\u001b[90me\u001b[39m7                41.4\n",
      "\u001b[90m 6\u001b[39m chr10:11005244:AG chr10:11… CELF2     intron 10    1.10\u001b[90me\u001b[39m7                16.8\n",
      "\u001b[90m 7\u001b[39m chr10:11032146:A  chr10:11… CELF2     intron 10    1.10\u001b[90me\u001b[39m7                31.9\n",
      "\u001b[90m 8\u001b[39m chr10:11086578:A  chr10:11… CELF2     intron 10    1.11\u001b[90me\u001b[39m7                \u001b[31mNA\u001b[39m  \n",
      "\u001b[90m 9\u001b[39m chr10:11103487:T  chr10:11… CELF2     intron 10    1.11\u001b[90me\u001b[39m7                47.0\n",
      "\u001b[90m10\u001b[39m chr10:11119864:G  chr10:11… CELF2     intron 10    1.11\u001b[90me\u001b[39m7                61.3\n",
      "\u001b[90m# ℹ 13 more variables: min_allele2_case <dbl>, max_allele2_case <dbl>,\u001b[39m\n",
      "\u001b[90m#   mean_allele2_case <dbl>, sd_allele2_case <dbl>, n_samples_case <int>,\u001b[39m\n",
      "\u001b[90m#   median_allele2_control <dbl>, min_allele2_control <dbl>,\u001b[39m\n",
      "\u001b[90m#   max_allele2_control <dbl>, mean_allele2_control <dbl>,\u001b[39m\n",
      "\u001b[90m#   sd_allele2_control <dbl>, n_samples_control <int>, has_no_overlap <lgl>,\u001b[39m\n",
      "\u001b[90m#   overlap_type <chr>\u001b[39m\n",
      "✓ Saved: allele2_no_overlap_statistics_per_locus.csv\n",
      "\n",
      "--- Creating No-Overlap Variants CSV ---\n",
      "\u001b[90m# A tibble: 3 × 15\u001b[39m\n",
      "  locus_definition STRs_ID     gene_name region chrom  start median_allele2_case\n",
      "  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m            \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m       \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m     \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m               \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m\n",
      "\u001b[90m1\u001b[39m chr3:41935209:T  chr3:41935… ULK4      intron 3     4.19\u001b[90me\u001b[39m7                54.0\n",
      "\u001b[90m2\u001b[39m chr12:77471132:A chr12:7747… NAV3      intron 12    7.75\u001b[90me\u001b[39m7                31.7\n",
      "\u001b[90m3\u001b[39m chr3:71057291:T  chr3:71057… FOXP1     intron 3     7.11\u001b[90me\u001b[39m7                25.8\n",
      "\u001b[90m# ℹ 8 more variables: min_allele2_case <dbl>, max_allele2_case <dbl>,\u001b[39m\n",
      "\u001b[90m#   n_samples_case <int>, median_allele2_control <dbl>,\u001b[39m\n",
      "\u001b[90m#   min_allele2_control <dbl>, max_allele2_control <dbl>,\u001b[39m\n",
      "\u001b[90m#   n_samples_control <int>, overlap_type <chr>\u001b[39m\n",
      "✓ Saved: allele2_no_overlap_variants_only.csv\n",
      "\n",
      "=== NO-OVERLAP ANALYSIS SUMMARY (Allele 2) ===\n",
      "Total loci analyzed: 4046 \n",
      "Loci with NO-OVERLAP: 3 \n",
      "\n",
      "No-overlap breakdown:\n",
      "\u001b[90m# A tibble: 2 × 2\u001b[39m\n",
      "  overlap_type   count\n",
      "  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m          \u001b[3m\u001b[90m<int>\u001b[39m\u001b[23m\n",
      "\u001b[90m1\u001b[39m case_higher        1\n",
      "\u001b[90m2\u001b[39m control_higher     2\n"
     ]
    }
   ],
   "source": [
    "# =======================================================\n",
    "# ANALYSIS: ALLELE 2 NO-OVERLAP DETECTION (CASE vs CONTROL)\n",
    "# =======================================================\n",
    "cat(\"--- Starting No-Overlap Analysis for Allele 2 ---\\n\")\n",
    "\n",
    "\n",
    "# =======================================================\n",
    "# 1. PREPARE DATA FOR NO-OVERLAP ANALYSIS\n",
    "# =======================================================\n",
    "\n",
    "# Force recreation with the correct metric (Allele 2)\n",
    "cat(\"--- Preparing data with Allele 2 metric ---\\n\")\n",
    "\n",
    "df_prepared <- df_filtered_grouped %>%\n",
    "  mutate(locus_definition = str_remove(STRs_ID, \":[^:]+$\")) %>%\n",
    "  mutate(allele2_val = allele2_est)\n",
    "\n",
    "df_per_locus <- df_prepared %>%\n",
    "  group_by(locus_definition, sample_id, group) %>%\n",
    "  summarise(\n",
    "    allele2_val = mean(allele2_val, na.rm = TRUE),\n",
    "    gene_name = first(gene_name),\n",
    "    chrom     = first(chrom),\n",
    "    start     = first(start),\n",
    "    region    = first(region),\n",
    "    STRs_ID   = first(STRs_ID),\n",
    "    .groups = \"drop\"\n",
    "  )\n",
    "\n",
    "cat(\"Data prepared. Rows:\", nrow(df_per_locus), \"\\n\")\n",
    "\n",
    "\n",
    "# =======================================================\n",
    "# 2. CALCULATE STATISTICS PER LOCUS AND GROUP\n",
    "# =======================================================\n",
    "cat(\"\\n--- Computing Statistics per Locus ---\\n\")\n",
    "\n",
    "locus_stats <- df_per_locus %>%\n",
    "  filter(!is.na(allele2_val)) %>%\n",
    "  group_by(locus_definition, group) %>%\n",
    "  summarise(\n",
    "    median_allele2 = median(allele2_val, na.rm = TRUE),\n",
    "    min_allele2    = min(allele2_val, na.rm = TRUE),\n",
    "    max_allele2    = max(allele2_val, na.rm = TRUE),\n",
    "    mean_allele2   = mean(allele2_val, na.rm = TRUE),\n",
    "    sd_allele2     = sd(allele2_val, na.rm = TRUE),\n",
    "    n_samples      = n(),\n",
    "    .groups = \"drop\"\n",
    "  )\n",
    "\n",
    "# Display the structure\n",
    "cat(\"Locus stats calculated. Rows:\", nrow(locus_stats), \"\\n\")\n",
    "print(head(locus_stats))\n",
    "\n",
    "\n",
    "# =======================================================\n",
    "# 3. IDENTIFY NO-OVERLAP VARIANTS\n",
    "# =======================================================\n",
    "cat(\"\\n--- Identifying No-Overlap Variants ---\\n\")\n",
    "\n",
    "# Pivot to compare case vs control side by side\n",
    "locus_comparison <- locus_stats %>%\n",
    "  pivot_wider(\n",
    "    id_cols = locus_definition,\n",
    "    names_from = group,\n",
    "    values_from = c(median_allele2, min_allele2, max_allele2, mean_allele2, sd_allele2, n_samples),\n",
    "    names_glue = \"{.value}_{group}\"\n",
    "  )\n",
    "\n",
    "# Detect no-overlap:\n",
    "# - If min_allele2_case > max_allele2_control: case all greater\n",
    "# - If min_allele2_control > max_allele2_case: control all greater\n",
    "MIN_SAMPLES_PER_GROUP <- 3\n",
    "\n",
    "locus_comparison <- locus_comparison %>%\n",
    "  mutate(\n",
    "    has_no_overlap = (n_samples_case >= MIN_SAMPLES_PER_GROUP & \n",
    "                      n_samples_control >= MIN_SAMPLES_PER_GROUP) &\n",
    "                     ((min_allele2_case > max_allele2_control) | \n",
    "                      (min_allele2_control > max_allele2_case)),\n",
    "    \n",
    "    overlap_type = case_when(\n",
    "      n_samples_case < MIN_SAMPLES_PER_GROUP ~ \"insufficient_case_samples\",\n",
    "      n_samples_control < MIN_SAMPLES_PER_GROUP ~ \"insufficient_control_samples\",\n",
    "      min_allele2_case > max_allele2_control ~ \"case_higher\",\n",
    "      min_allele2_control > max_allele2_case ~ \"control_higher\",\n",
    "      TRUE ~ \"overlap\"\n",
    "    )\n",
    "  )\n",
    "\n",
    "# Recover metadata\n",
    "locus_with_metadata <- locus_comparison %>%\n",
    "  left_join(\n",
    "    df_per_locus %>% \n",
    "      select(locus_definition, gene_name, region, STRs_ID, chrom, start) %>%\n",
    "      distinct(),\n",
    "    by = \"locus_definition\"\n",
    "  )\n",
    "\n",
    "cat(\"Total loci:\", nrow(locus_with_metadata), \"\\n\")\n",
    "cat(\"Loci with NO-OVERLAP (min\", MIN_SAMPLES_PER_GROUP, \"samples per group):\", \n",
    "    sum(locus_with_metadata$has_no_overlap, na.rm = TRUE), \"\\n\")\n",
    "cat(\"\\n\")\n",
    "cat(\"Breakdown by category:\\n\")\n",
    "cat(\"  case_higher:\", sum(locus_with_metadata$overlap_type == \"case_higher\", na.rm = TRUE), \"\\n\")\n",
    "cat(\"  control_higher:\", sum(locus_with_metadata$overlap_type == \"control_higher\", na.rm = TRUE), \"\\n\")\n",
    "cat(\"  insufficient_case_samples:\", sum(locus_with_metadata$overlap_type == \"insufficient_case_samples\", na.rm = TRUE), \"\\n\")\n",
    "cat(\"  insufficient_control_samples:\", sum(locus_with_metadata$overlap_type == \"insufficient_control_samples\", na.rm = TRUE), \"\\n\")\n",
    "cat(\"  overlap:\", sum(locus_with_metadata$overlap_type == \"overlap\", na.rm = TRUE), \"\\n\")\n",
    "cat(\"\\n\")\n",
    "\n",
    "\n",
    "# =======================================================\n",
    "# 4. CREATE OUTPUT CSV 1: DETAILED STATISTICS PER LOCUS\n",
    "# =======================================================\n",
    "cat(\"--- Creating Statistics CSV ---\\n\")\n",
    "\n",
    "csv_statistics <- locus_with_metadata %>%\n",
    "  select(\n",
    "    locus_definition,\n",
    "    STRs_ID,\n",
    "    gene_name,\n",
    "    region,\n",
    "    chrom,\n",
    "    start,\n",
    "    \n",
    "    # Case group\n",
    "    median_allele2_case,\n",
    "    min_allele2_case,\n",
    "    max_allele2_case,\n",
    "    mean_allele2_case,\n",
    "    sd_allele2_case,\n",
    "    n_samples_case,\n",
    "    \n",
    "    # Control group\n",
    "    median_allele2_control,\n",
    "    min_allele2_control,\n",
    "    max_allele2_control,\n",
    "    mean_allele2_control,\n",
    "    sd_allele2_control,\n",
    "    n_samples_control,\n",
    "    \n",
    "    # Overlap info\n",
    "    has_no_overlap,\n",
    "    overlap_type\n",
    "  ) %>%\n",
    "  arrange(desc(has_no_overlap), locus_definition)\n",
    "\n",
    "print(head(csv_statistics, 10))\n",
    "\n",
    "# Create results directory if it doesn't exist\n",
    "if(!dir.exists(\"results\")) dir.create(\"results\")\n",
    "\n",
    "write_csv(\n",
    "  csv_statistics,\n",
    "  \"results/allele2_no_overlap_statistics_per_locus.csv\"\n",
    ")\n",
    "\n",
    "cat(\"✓ Saved: allele2_no_overlap_statistics_per_locus.csv\\n\")\n",
    "\n",
    "\n",
    "# =======================================================\n",
    "# 5. CREATE OUTPUT CSV 2: ONLY NO-OVERLAP VARIANTS\n",
    "# =======================================================\n",
    "cat(\"\\n--- Creating No-Overlap Variants CSV ---\\n\")\n",
    "\n",
    "csv_no_overlap_only <- locus_with_metadata %>%\n",
    "  filter(has_no_overlap == TRUE) %>%\n",
    "  select(\n",
    "    locus_definition,\n",
    "    STRs_ID,\n",
    "    gene_name,\n",
    "    region,\n",
    "    chrom,\n",
    "    start,\n",
    "    \n",
    "    # Case group\n",
    "    median_allele2_case,\n",
    "    min_allele2_case,\n",
    "    max_allele2_case,\n",
    "    n_samples_case,\n",
    "    \n",
    "    # Control group\n",
    "    median_allele2_control,\n",
    "    min_allele2_control,\n",
    "    max_allele2_control,\n",
    "    n_samples_control,\n",
    "    \n",
    "    # Interpretation\n",
    "    overlap_type\n",
    "  ) %>%\n",
    "  arrange(overlap_type, locus_definition)\n",
    "\n",
    "print(csv_no_overlap_only)\n",
    "\n",
    "write_csv(\n",
    "  csv_no_overlap_only,\n",
    "  \"results/allele2_no_overlap_variants_only.csv\"\n",
    ")\n",
    "\n",
    "cat(\"✓ Saved: allele2_no_overlap_variants_only.csv\\n\")\n",
    "\n",
    "\n",
    "# =======================================================\n",
    "# 6. SUMMARY REPORT\n",
    "# =======================================================\n",
    "cat(\"\\n=== NO-OVERLAP ANALYSIS SUMMARY (Allele 2) ===\\n\")\n",
    "cat(\"Total loci analyzed:\", nrow(csv_statistics), \"\\n\")\n",
    "cat(\"Loci with NO-OVERLAP:\", nrow(csv_no_overlap_only), \"\\n\")\n",
    "cat(\"\\nNo-overlap breakdown:\\n\")\n",
    "\n",
    "if(nrow(csv_no_overlap_only) > 0) {\n",
    "  breakdown <- csv_no_overlap_only %>%\n",
    "    group_by(overlap_type) %>%\n",
    "    summarise(\n",
    "      count = n(),\n",
    "      .groups = \"drop\"\n",
    "    )\n",
    "  \n",
    "  print(breakdown)\n",
    "} else {\n",
    "  cat(\"No variants without overlap found.\\n\")\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8fbceb79",
   "metadata": {},
   "source": [
    "6. Identification of loci without overlap in relation of mean alleles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "6a80c87f",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--- Starting No-Overlap Analysis for Mean Alleles ---\n",
      "--- Preparing data with Mean Alleles metric ---\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Data prepared. Rows: 144796 \n",
      "\n",
      "--- Computing Statistics per Locus ---\n",
      "Locus stats calculated. Rows: 7447 \n",
      "\u001b[90m# A tibble: 6 × 8\u001b[39m\n",
      "  locus_definition  group   median_mean min_mean max_mean mean_value sd_mean\n",
      "  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m             \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m         \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m    \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m    \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m      \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m   \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m\n",
      "\u001b[90m1\u001b[39m chr10:10857639:AT case           8.43     5.78     12.2       8.91    2.13\n",
      "\u001b[90m2\u001b[39m chr10:10857639:AT control        9.49     2.86     19.0      10.1     3.34\n",
      "\u001b[90m3\u001b[39m chr10:10994724:C  case          20.7     14.8      31.0      21.1     5.09\n",
      "\u001b[90m4\u001b[39m chr10:10994724:C  control       17.1     13.5      30.5      18.7     3.84\n",
      "\u001b[90m5\u001b[39m chr10:11005244:AG case           8.13     5.96     14.6       8.75    2.78\n",
      "\u001b[90m6\u001b[39m chr10:11005244:AG control       10.1      2.33     18.5       9.81    3.55\n",
      "\u001b[90m# ℹ 1 more variable: n_samples <int>\u001b[39m\n",
      "\n",
      "--- Identifying No-Overlap Variants ---\n",
      "Total loci: 4046 \n",
      "Loci with NO-OVERLAP (min 3 samples per group): 2 \n",
      "\n",
      "--- Creating Statistics CSV ---\n",
      "\u001b[90m# A tibble: 10 × 20\u001b[39m\n",
      "   locus_definition  STRs_ID      gene_name region chrom  start median_mean_case\n",
      "   \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m             \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m        \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m     \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m            \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m\n",
      "\u001b[90m 1\u001b[39m chr12:77471132:A  chr12:77471… NAV3      intron 12    7.75\u001b[90me\u001b[39m7            14.9 \n",
      "\u001b[90m 2\u001b[39m chr3:71057291:T   chr3:710572… FOXP1     intron 3     7.11\u001b[90me\u001b[39m7            14.3 \n",
      "\u001b[90m 3\u001b[39m chr10:10857639:AT chr10:10857… CELF2     intron 10    1.09\u001b[90me\u001b[39m7             8.43\n",
      "\u001b[90m 4\u001b[39m chr10:10994724:C  chr10:10994… CELF2     intron 10    1.10\u001b[90me\u001b[39m7            20.7 \n",
      "\u001b[90m 5\u001b[39m chr10:11005244:AG chr10:11005… CELF2     intron 10    1.10\u001b[90me\u001b[39m7             8.13\n",
      "\u001b[90m 6\u001b[39m chr10:11032146:A  chr10:11032… CELF2     intron 10    1.10\u001b[90me\u001b[39m7            14.4 \n",
      "\u001b[90m 7\u001b[39m chr10:11086578:A  chr10:11086… CELF2     intron 10    1.11\u001b[90me\u001b[39m7            \u001b[31mNA\u001b[39m   \n",
      "\u001b[90m 8\u001b[39m chr10:11103487:T  chr10:11103… CELF2     intron 10    1.11\u001b[90me\u001b[39m7            23.5 \n",
      "\u001b[90m 9\u001b[39m chr10:11119864:G  chr10:11119… CELF2     intron 10    1.11\u001b[90me\u001b[39m7            30.7 \n",
      "\u001b[90m10\u001b[39m chr10:11165056:C  chr10:11165… CELF2     five_… 10    1.12\u001b[90me\u001b[39m7            30.9 \n",
      "\u001b[90m# ℹ 13 more variables: min_mean_case <dbl>, max_mean_case <dbl>,\u001b[39m\n",
      "\u001b[90m#   mean_value_case <dbl>, sd_mean_case <dbl>, n_samples_case <int>,\u001b[39m\n",
      "\u001b[90m#   median_mean_control <dbl>, min_mean_control <dbl>, max_mean_control <dbl>,\u001b[39m\n",
      "\u001b[90m#   mean_value_control <dbl>, sd_mean_control <dbl>, n_samples_control <int>,\u001b[39m\n",
      "\u001b[90m#   has_no_overlap <lgl>, overlap_type <chr>\u001b[39m\n",
      "✓ Saved: mean_alleles_no_overlap_statistics_per_locus.csv\n",
      "\n",
      "--- Creating No-Overlap Variants CSV ---\n",
      "\u001b[90m# A tibble: 2 × 15\u001b[39m\n",
      "  locus_definition STRs_ID        gene_name region chrom  start median_mean_case\n",
      "  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m            \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m          \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m     \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m            \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m\n",
      "\u001b[90m1\u001b[39m chr12:77471132:A chr12:7747113… NAV3      intron 12    7.75\u001b[90me\u001b[39m7             14.9\n",
      "\u001b[90m2\u001b[39m chr3:71057291:T  chr3:71057291… FOXP1     intron 3     7.11\u001b[90me\u001b[39m7             14.3\n",
      "\u001b[90m# ℹ 8 more variables: min_mean_case <dbl>, max_mean_case <dbl>,\u001b[39m\n",
      "\u001b[90m#   n_samples_case <int>, median_mean_control <dbl>, min_mean_control <dbl>,\u001b[39m\n",
      "\u001b[90m#   max_mean_control <dbl>, n_samples_control <int>, overlap_type <chr>\u001b[39m\n",
      "✓ Saved: mean_alleles_no_overlap_variants_only.csv\n",
      "\n",
      "=== NO-OVERLAP ANALYSIS SUMMARY (Mean Alleles) ===\n",
      "Total loci analyzed: 4046 \n",
      "Loci with NO-OVERLAP (min 3 samples per group): 2 \n",
      "\n",
      "No-overlap breakdown:\n",
      "\u001b[90m# A tibble: 1 × 2\u001b[39m\n",
      "  overlap_type   count\n",
      "  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m          \u001b[3m\u001b[90m<int>\u001b[39m\u001b[23m\n",
      "\u001b[90m1\u001b[39m control_higher     2\n"
     ]
    }
   ],
   "source": [
    "# =======================================================\n",
    "# ANALYSIS: MEAN ALLELES NO-OVERLAP DETECTION (CASE vs CONTROL)\n",
    "# =======================================================\n",
    "cat(\"--- Starting No-Overlap Analysis for Mean Alleles ---\\n\")\n",
    "\n",
    "\n",
    "# =======================================================\n",
    "# 1. PREPARE DATA FOR NO-OVERLAP ANALYSIS\n",
    "# =======================================================\n",
    "\n",
    "# Force recreation with the correct metric (Mean Alleles)\n",
    "cat(\"--- Preparing data with Mean Alleles metric ---\\n\")\n",
    "\n",
    "df_prepared <- df_filtered_grouped %>%\n",
    "  mutate(locus_definition = str_remove(STRs_ID, \":[^:]+$\")) %>%\n",
    "  mutate(mean_alleles_val = (allele1_est + allele2_est) / 2)\n",
    "\n",
    "df_per_locus <- df_prepared %>%\n",
    "  group_by(locus_definition, sample_id, group) %>%\n",
    "  summarise(\n",
    "    mean_alleles_val = mean(mean_alleles_val, na.rm = TRUE),\n",
    "    gene_name = first(gene_name),\n",
    "    chrom     = first(chrom),\n",
    "    start     = first(start),\n",
    "    region    = first(region),\n",
    "    STRs_ID   = first(STRs_ID),\n",
    "    .groups = \"drop\"\n",
    "  )\n",
    "\n",
    "cat(\"Data prepared. Rows:\", nrow(df_per_locus), \"\\n\")\n",
    "\n",
    "\n",
    "# =======================================================\n",
    "# 2. CALCULATE STATISTICS PER LOCUS AND GROUP\n",
    "# =======================================================\n",
    "cat(\"\\n--- Computing Statistics per Locus ---\\n\")\n",
    "\n",
    "locus_stats <- df_per_locus %>%\n",
    "  filter(!is.na(mean_alleles_val)) %>%\n",
    "  group_by(locus_definition, group) %>%\n",
    "  summarise(\n",
    "    median_mean = median(mean_alleles_val, na.rm = TRUE),\n",
    "    min_mean    = min(mean_alleles_val, na.rm = TRUE),\n",
    "    max_mean    = max(mean_alleles_val, na.rm = TRUE),\n",
    "    mean_value  = mean(mean_alleles_val, na.rm = TRUE),\n",
    "    sd_mean     = sd(mean_alleles_val, na.rm = TRUE),\n",
    "    n_samples   = n(),\n",
    "    .groups = \"drop\"\n",
    "  )\n",
    "\n",
    "# Display the structure\n",
    "cat(\"Locus stats calculated. Rows:\", nrow(locus_stats), \"\\n\")\n",
    "print(head(locus_stats))\n",
    "\n",
    "\n",
    "# =======================================================\n",
    "# 3. IDENTIFY NO-OVERLAP VARIANTS\n",
    "# =======================================================\n",
    "cat(\"\\n--- Identifying No-Overlap Variants ---\\n\")\n",
    "\n",
    "# Pivot to compare case vs control side by side\n",
    "locus_comparison <- locus_stats %>%\n",
    "  pivot_wider(\n",
    "    id_cols = locus_definition,\n",
    "    names_from = group,\n",
    "    values_from = c(median_mean, min_mean, max_mean, mean_value, sd_mean, n_samples),\n",
    "    names_glue = \"{.value}_{group}\"\n",
    "  )\n",
    "\n",
    "# Detect no-overlap WITH MINIMUM SAMPLE SIZE REQUIREMENT\n",
    "# Only consider variants with at least N samples in BOTH groups\n",
    "MIN_SAMPLES_PER_GROUP <- 3  # Adjust this threshold as needed\n",
    "\n",
    "locus_comparison <- locus_comparison %>%\n",
    "  mutate(\n",
    "    has_no_overlap = (n_samples_case >= MIN_SAMPLES_PER_GROUP & \n",
    "                      n_samples_control >= MIN_SAMPLES_PER_GROUP) &\n",
    "                     ((min_mean_case > max_mean_control) | \n",
    "                      (min_mean_control > max_mean_case)),\n",
    "    \n",
    "    overlap_type = case_when(\n",
    "      n_samples_case < MIN_SAMPLES_PER_GROUP ~ \"insufficient_case_samples\",\n",
    "      n_samples_control < MIN_SAMPLES_PER_GROUP ~ \"insufficient_control_samples\",\n",
    "      min_mean_case > max_mean_control ~ \"case_higher\",\n",
    "      min_mean_control > max_mean_case ~ \"control_higher\",\n",
    "      TRUE ~ \"overlap\"\n",
    "    )\n",
    "  )\n",
    "\n",
    "# Recover metadata\n",
    "locus_with_metadata <- locus_comparison %>%\n",
    "  left_join(\n",
    "    df_per_locus %>% \n",
    "      select(locus_definition, gene_name, region, STRs_ID, chrom, start) %>%\n",
    "      distinct(),\n",
    "    by = \"locus_definition\"\n",
    "  )\n",
    "\n",
    "cat(\"Total loci:\", nrow(locus_with_metadata), \"\\n\")\n",
    "cat(\"Loci with NO-OVERLAP (min\", MIN_SAMPLES_PER_GROUP, \"samples per group):\", \n",
    "    sum(locus_with_metadata$has_no_overlap, na.rm = TRUE), \"\\n\")\n",
    "cat(\"\\n\")\n",
    "\n",
    "\n",
    "# =======================================================\n",
    "# 4. CREATE OUTPUT CSV 1: DETAILED STATISTICS PER LOCUS\n",
    "# =======================================================\n",
    "cat(\"--- Creating Statistics CSV ---\\n\")\n",
    "\n",
    "csv_statistics <- locus_with_metadata %>%\n",
    "  select(\n",
    "    locus_definition,\n",
    "    STRs_ID,\n",
    "    gene_name,\n",
    "    region,\n",
    "    chrom,\n",
    "    start,\n",
    "    \n",
    "    # Case group\n",
    "    median_mean_case,\n",
    "    min_mean_case,\n",
    "    max_mean_case,\n",
    "    mean_value_case,\n",
    "    sd_mean_case,\n",
    "    n_samples_case,\n",
    "    \n",
    "    # Control group\n",
    "    median_mean_control,\n",
    "    min_mean_control,\n",
    "    max_mean_control,\n",
    "    mean_value_control,\n",
    "    sd_mean_control,\n",
    "    n_samples_control,\n",
    "    \n",
    "    # Overlap info\n",
    "    has_no_overlap,\n",
    "    overlap_type\n",
    "  ) %>%\n",
    "  arrange(desc(has_no_overlap), locus_definition)\n",
    "\n",
    "print(head(csv_statistics, 10))\n",
    "\n",
    "# Create results directory if it doesn't exist\n",
    "if(!dir.exists(\"results\")) dir.create(\"results\")\n",
    "\n",
    "write_csv(\n",
    "  csv_statistics,\n",
    "  \"results/mean_alleles_no_overlap_statistics_per_locus.csv\"\n",
    ")\n",
    "\n",
    "cat(\"✓ Saved: mean_alleles_no_overlap_statistics_per_locus.csv\\n\")\n",
    "\n",
    "\n",
    "# =======================================================\n",
    "# 5. CREATE OUTPUT CSV 2: ONLY NO-OVERLAP VARIANTS\n",
    "# =======================================================\n",
    "cat(\"\\n--- Creating No-Overlap Variants CSV ---\\n\")\n",
    "\n",
    "csv_no_overlap_only <- locus_with_metadata %>%\n",
    "  filter(has_no_overlap == TRUE) %>%\n",
    "  select(\n",
    "    locus_definition,\n",
    "    STRs_ID,\n",
    "    gene_name,\n",
    "    region,\n",
    "    chrom,\n",
    "    start,\n",
    "    \n",
    "    # Case group\n",
    "    median_mean_case,\n",
    "    min_mean_case,\n",
    "    max_mean_case,\n",
    "    n_samples_case,\n",
    "    \n",
    "    # Control group\n",
    "    median_mean_control,\n",
    "    min_mean_control,\n",
    "    max_mean_control,\n",
    "    n_samples_control,\n",
    "    \n",
    "    # Interpretation\n",
    "    overlap_type\n",
    "  ) %>%\n",
    "  arrange(overlap_type, locus_definition)\n",
    "\n",
    "print(csv_no_overlap_only)\n",
    "\n",
    "write_csv(\n",
    "  csv_no_overlap_only,\n",
    "  \"results/mean_alleles_no_overlap_variants_only.csv\"\n",
    ")\n",
    "\n",
    "cat(\"✓ Saved: mean_alleles_no_overlap_variants_only.csv\\n\")\n",
    "\n",
    "\n",
    "# =======================================================\n",
    "# 6. SUMMARY REPORT\n",
    "# =======================================================\n",
    "cat(\"\\n=== NO-OVERLAP ANALYSIS SUMMARY (Mean Alleles) ===\\n\")\n",
    "cat(\"Total loci analyzed:\", nrow(csv_statistics), \"\\n\")\n",
    "cat(\"Loci with NO-OVERLAP (min\", MIN_SAMPLES_PER_GROUP, \"samples per group):\", \n",
    "    nrow(csv_no_overlap_only), \"\\n\")\n",
    "cat(\"\\nNo-overlap breakdown:\\n\")\n",
    "\n",
    "if(nrow(csv_no_overlap_only) > 0) {\n",
    "  breakdown <- csv_no_overlap_only %>%\n",
    "    group_by(overlap_type) %>%\n",
    "    summarise(\n",
    "      count = n(),\n",
    "      .groups = \"drop\"\n",
    "    )\n",
    "  \n",
    "  print(breakdown)\n",
    "} else {\n",
    "  cat(\"No variants without overlap found.\\n\")\n",
    "}"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.5.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
