{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "c5ba2d8d",
   "metadata": {},
   "source": [
    "# Preparing the environment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "e5f9ef20",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: regioneR\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    findMatches\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "Loading required package: BSgenome\n",
      "\n",
      "Loading required package: Biostrings\n",
      "\n",
      "Loading required package: XVector\n",
      "\n",
      "\n",
      "Attaching package: 'Biostrings'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:base':\n",
      "\n",
      "    strsplit\n",
      "\n",
      "\n",
      "Loading required package: BiocIO\n",
      "\n",
      "Loading required package: rtracklayer\n",
      "\n",
      "\n",
      "Attaching package: 'rtracklayer'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:BiocIO':\n",
      "\n",
      "    FileForFormat\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: 'dplyr'\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:Biostrings':\n",
      "\n",
      "    collapse, intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "The following object is masked from 'package:XVector':\n",
      "\n",
      "    slice\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:GenomicRanges':\n",
      "\n",
      "    intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from 'package:GenomeInfoDb':\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:IRanges':\n",
      "\n",
      "    collapse, desc, intersect, setdiff, slice, union\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:S4Vectors':\n",
      "\n",
      "    first, intersect, rename, setdiff, setequal, union\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:BiocGenerics':\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:stats':\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:base':\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "========================================\n",
      "circlize version 0.4.16\n",
      "CRAN page: https://cran.r-project.org/package=circlize\n",
      "Github page: https://github.com/jokergoo/circlize\n",
      "Documentation: https://jokergoo.github.io/circlize_book/book/\n",
      "\n",
      "If you use it in published research, please cite:\n",
      "Gu, Z. circlize implements and enhances circular visualization\n",
      "  in R. Bioinformatics 2014.\n",
      "\n",
      "This message can be suppressed by:\n",
      "  suppressPackageStartupMessages(library(circlize))\n",
      "========================================\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(karyoploteR)\n",
    "library(GenomicRanges)\n",
    "library(BSgenome.Hsapiens.UCSC.hg38)\n",
    "library(ggplot2)\n",
    "library(dplyr)\n",
    "library(circlize)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8743a9c7",
   "metadata": {},
   "source": [
    "# Importing files for use with KaryoploteR"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d10e8568",
   "metadata": {},
   "source": [
    "Data import"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "0b52c2c9",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 17</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>STRs_ID</th><th scope=col>region</th><th scope=col>gene_id</th><th scope=col>priority</th><th scope=col>gene_name</th><th scope=col>gene_chrom</th><th scope=col>gene_start</th><th scope=col>gene_end</th><th scope=col>annotation</th><th scope=col>sample_id</th><th scope=col>chrom</th><th scope=col>start</th><th scope=col>end</th><th scope=col>repeat_unit</th><th scope=col>allele1_est</th><th scope=col>allele2_est</th><th scope=col>depth</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>chr16:12744297:AAGC:12 </td><td>promoter  </td><td>ENSG00000260378</td><td>4</td><td>AC109597.1</td><td>chr16</td><td>12745873</td><td>12757835</td><td>promoter:AC109597.1</td><td>BEL-1012-1</td><td>16</td><td>12744297</td><td>12744345</td><td>AAGC  </td><td>-0.50</td><td>0.00</td><td>36</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>chr17:4413485:CTTG:11  </td><td>intergenic</td><td>intergenic     </td><td>6</td><td>.         </td><td>.    </td><td>      -1</td><td>      -1</td><td>intergenic         </td><td>BEL-1012-1</td><td>17</td><td> 4413485</td><td> 4413530</td><td>CTTG  </td><td> 5.00</td><td>5.73</td><td>25</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>chr20:62334753:AGGGCG:5</td><td>intron    </td><td>ENSG00000130702</td><td>5</td><td>LAMA5     </td><td>chr20</td><td>62307955</td><td>62367312</td><td>intron:LAMA5       </td><td>BEL-1012-1</td><td>20</td><td>62334753</td><td>62334783</td><td>AGGGCG</td><td> 3.17</td><td>0.00</td><td>46</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>chr22:32780933:CCTT:0  </td><td>promoter  </td><td>ENSG00000229673</td><td>4</td><td>Z73495.1  </td><td>chr22</td><td>32783177</td><td>32784982</td><td>promoter:Z73495.1  </td><td>BEL-1012-1</td><td>22</td><td>32780933</td><td>32780934</td><td>CCTT  </td><td>-1.00</td><td>0.00</td><td>30</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>chr20:41707548:AAGG:0  </td><td>intergenic</td><td>intergenic     </td><td>6</td><td>.         </td><td>.    </td><td>      -1</td><td>      -1</td><td>intergenic         </td><td>BEL-1012-1</td><td>20</td><td>41707548</td><td>41707549</td><td>AAGG  </td><td> 0.00</td><td>0.00</td><td>28</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>chr7:92128476:GT:2     </td><td>intron    </td><td>ENSG00000001630</td><td>5</td><td>CYP51A1   </td><td>chr7 </td><td>92112153</td><td>92134803</td><td>intron:CYP51A1     </td><td>BEL-1012-1</td><td>7 </td><td>92128476</td><td>92128479</td><td>GT    </td><td> 0.00</td><td>0.00</td><td>40</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 17\n",
       "\\begin{tabular}{r|lllllllllllllllll}\n",
       "  & STRs\\_ID & region & gene\\_id & priority & gene\\_name & gene\\_chrom & gene\\_start & gene\\_end & annotation & sample\\_id & chrom & start & end & repeat\\_unit & allele1\\_est & allele2\\_est & depth\\\\\n",
       "  & <chr> & <chr> & <chr> & <int> & <chr> & <chr> & <int> & <int> & <chr> & <chr> & <chr> & <int> & <int> & <chr> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t1 & chr16:12744297:AAGC:12  & promoter   & ENSG00000260378 & 4 & AC109597.1 & chr16 & 12745873 & 12757835 & promoter:AC109597.1 & BEL-1012-1 & 16 & 12744297 & 12744345 & AAGC   & -0.50 & 0.00 & 36\\\\\n",
       "\t2 & chr17:4413485:CTTG:11   & intergenic & intergenic      & 6 & .          & .     &       -1 &       -1 & intergenic          & BEL-1012-1 & 17 &  4413485 &  4413530 & CTTG   &  5.00 & 5.73 & 25\\\\\n",
       "\t3 & chr20:62334753:AGGGCG:5 & intron     & ENSG00000130702 & 5 & LAMA5      & chr20 & 62307955 & 62367312 & intron:LAMA5        & BEL-1012-1 & 20 & 62334753 & 62334783 & AGGGCG &  3.17 & 0.00 & 46\\\\\n",
       "\t4 & chr22:32780933:CCTT:0   & promoter   & ENSG00000229673 & 4 & Z73495.1   & chr22 & 32783177 & 32784982 & promoter:Z73495.1   & BEL-1012-1 & 22 & 32780933 & 32780934 & CCTT   & -1.00 & 0.00 & 30\\\\\n",
       "\t5 & chr20:41707548:AAGG:0   & intergenic & intergenic      & 6 & .          & .     &       -1 &       -1 & intergenic          & BEL-1012-1 & 20 & 41707548 & 41707549 & AAGG   &  0.00 & 0.00 & 28\\\\\n",
       "\t6 & chr7:92128476:GT:2      & intron     & ENSG00000001630 & 5 & CYP51A1    & chr7  & 92112153 & 92134803 & intron:CYP51A1      & BEL-1012-1 & 7  & 92128476 & 92128479 & GT     &  0.00 & 0.00 & 40\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 17\n",
       "\n",
       "| <!--/--> | STRs_ID &lt;chr&gt; | region &lt;chr&gt; | gene_id &lt;chr&gt; | priority &lt;int&gt; | gene_name &lt;chr&gt; | gene_chrom &lt;chr&gt; | gene_start &lt;int&gt; | gene_end &lt;int&gt; | annotation &lt;chr&gt; | sample_id &lt;chr&gt; | chrom &lt;chr&gt; | start &lt;int&gt; | end &lt;int&gt; | repeat_unit &lt;chr&gt; | allele1_est &lt;dbl&gt; | allele2_est &lt;dbl&gt; | depth &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | chr16:12744297:AAGC:12  | promoter   | ENSG00000260378 | 4 | AC109597.1 | chr16 | 12745873 | 12757835 | promoter:AC109597.1 | BEL-1012-1 | 16 | 12744297 | 12744345 | AAGC   | -0.50 | 0.00 | 36 |\n",
       "| 2 | chr17:4413485:CTTG:11   | intergenic | intergenic      | 6 | .          | .     |       -1 |       -1 | intergenic          | BEL-1012-1 | 17 |  4413485 |  4413530 | CTTG   |  5.00 | 5.73 | 25 |\n",
       "| 3 | chr20:62334753:AGGGCG:5 | intron     | ENSG00000130702 | 5 | LAMA5      | chr20 | 62307955 | 62367312 | intron:LAMA5        | BEL-1012-1 | 20 | 62334753 | 62334783 | AGGGCG |  3.17 | 0.00 | 46 |\n",
       "| 4 | chr22:32780933:CCTT:0   | promoter   | ENSG00000229673 | 4 | Z73495.1   | chr22 | 32783177 | 32784982 | promoter:Z73495.1   | BEL-1012-1 | 22 | 32780933 | 32780934 | CCTT   | -1.00 | 0.00 | 30 |\n",
       "| 5 | chr20:41707548:AAGG:0   | intergenic | intergenic      | 6 | .          | .     |       -1 |       -1 | intergenic          | BEL-1012-1 | 20 | 41707548 | 41707549 | AAGG   |  0.00 | 0.00 | 28 |\n",
       "| 6 | chr7:92128476:GT:2      | intron     | ENSG00000001630 | 5 | CYP51A1    | chr7  | 92112153 | 92134803 | intron:CYP51A1      | BEL-1012-1 | 7  | 92128476 | 92128479 | GT     |  0.00 | 0.00 | 40 |\n",
       "\n"
      ],
      "text/plain": [
       "  STRs_ID                 region     gene_id         priority gene_name \n",
       "1 chr16:12744297:AAGC:12  promoter   ENSG00000260378 4        AC109597.1\n",
       "2 chr17:4413485:CTTG:11   intergenic intergenic      6        .         \n",
       "3 chr20:62334753:AGGGCG:5 intron     ENSG00000130702 5        LAMA5     \n",
       "4 chr22:32780933:CCTT:0   promoter   ENSG00000229673 4        Z73495.1  \n",
       "5 chr20:41707548:AAGG:0   intergenic intergenic      6        .         \n",
       "6 chr7:92128476:GT:2      intron     ENSG00000001630 5        CYP51A1   \n",
       "  gene_chrom gene_start gene_end annotation          sample_id  chrom start   \n",
       "1 chr16      12745873   12757835 promoter:AC109597.1 BEL-1012-1 16    12744297\n",
       "2 .                -1         -1 intergenic          BEL-1012-1 17     4413485\n",
       "3 chr20      62307955   62367312 intron:LAMA5        BEL-1012-1 20    62334753\n",
       "4 chr22      32783177   32784982 promoter:Z73495.1   BEL-1012-1 22    32780933\n",
       "5 .                -1         -1 intergenic          BEL-1012-1 20    41707548\n",
       "6 chr7       92112153   92134803 intron:CYP51A1      BEL-1012-1 7     92128476\n",
       "  end      repeat_unit allele1_est allele2_est depth\n",
       "1 12744345 AAGC        -0.50       0.00        36   \n",
       "2  4413530 CTTG         5.00       5.73        25   \n",
       "3 62334783 AGGGCG       3.17       0.00        46   \n",
       "4 32780934 CCTT        -1.00       0.00        30   \n",
       "5 41707549 AAGG         0.00       0.00        28   \n",
       "6 92128479 GT           0.00       0.00        40   "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# File path\n",
    "original_file <- \"../samples/gp_global/merged/STRs_annotated_complete.tsv\"\n",
    "\n",
    "# Read cleaned file with read.delim\n",
    "df_annot <- read.delim(original_file, header=TRUE, sep=\"\\t\", stringsAsFactors = FALSE)\n",
    "\n",
    "# Show the first lines of the updated data frame\n",
    "head(df_annot)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0ae5506d",
   "metadata": {},
   "source": [
    "Quantidade de STRs a serem visualizadas e cromossomos disponíveis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "23129a69",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "STRs to plot: 14913155 \n",
      "\n",
      "            CDS  five_prime_utr      intergenic          intron          others \n",
      "          36290           61998         5427454         8185887           82992 \n",
      "       promoter three_prime_utr \n",
      "         960444          158090 \n",
      "\n",
      "Chromosomes available for plotting:\n",
      " [1] chr16 chr17 chr20 chr22 chr7  chr14 chr6  chr4  chr3  chr11 chr1  chr19\n",
      "[13] chr8  chr10 chrY  chr5  chr9  chr2  chr18 chr15 chr12 chr13 chr21 chrX \n",
      "24 Levels: chr16 chr17 chr20 chr22 chr7 chr14 chr6 chr4 chr3 chr11 ... chrX\n"
     ]
    }
   ],
   "source": [
    "# 1. Add \"chr\" prefix directly to the chrom column\n",
    "df_annot <- df_annot %>%\n",
    "  mutate(chrom = paste0(\"chr\", chrom))\n",
    "\n",
    "# 2. Define chromosomes with \"chr\" prefix\n",
    "chromosomes_to_plot <- c(paste0(\"chr\", 1:22), \"chrX\", \"chrY\")\n",
    "\n",
    "# 3. Filter\n",
    "df_filtered <- df_annot %>%\n",
    "  filter(chrom %in% chromosomes_to_plot)\n",
    "\n",
    "# 4. Create GRanges (already with correct chromosome names)\n",
    "gr_strs <- GRanges(\n",
    "  seqnames = df_filtered$chrom,\n",
    "  ranges = IRanges(start = df_filtered$start, end = df_filtered$end)\n",
    ")\n",
    "\n",
    "# 5. Add metadata\n",
    "mcols(gr_strs)$ID <- df_filtered$STRs_ID\n",
    "mcols(gr_strs)$Region <- df_filtered$region\n",
    "mcols(gr_strs)$Gene_Name <- df_filtered$gene_name\n",
    "\n",
    "# 6. Verify\n",
    "cat(\"STRs to plot:\", length(gr_strs), \"\\n\")\n",
    "print(table(gr_strs$Region))\n",
    "cat(\"\\nChromosomes available for plotting:\\n\")\n",
    "print(unique(seqnames(gr_strs)))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b1878602",
   "metadata": {},
   "source": [
    "# Plots"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "99b5e0c6",
   "metadata": {},
   "source": [
    "Plot simples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "ca441fb5",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Total STRs to plot: 14913155\"\n",
      "< table of extent 0 >\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<strong>agg_record_1584876767:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{agg\\textbackslash{}\\_record\\textbackslash{}\\_1584876767:} 2"
      ],
      "text/markdown": [
       "**agg_record_1584876767:** 2"
      ],
      "text/plain": [
       "agg_record_1584876767 \n",
       "                    2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Plot saved to: results/karyotype_STRs_distribution.png\n"
     ]
    }
   ],
   "source": [
    "# Create results directory if it doesn't exist\n",
    "if (!dir.exists(\"results\")) {\n",
    "  dir.create(\"results\")\n",
    "  cat(\"Created 'results' directory\\n\")\n",
    "}\n",
    "\n",
    "# Verify data\n",
    "print(paste(\"Total STRs to plot:\", length(gr_strs)))\n",
    "print(table(gr_strs$region))\n",
    "\n",
    "# ========================================\n",
    "# Option 1: Simple plot\n",
    "# ========================================\n",
    "\n",
    "# Open PNG device\n",
    "png(filename = \"results/karyotype_STRs_distribution.png\", \n",
    "    width = 3000, \n",
    "    height = 2000, \n",
    "    res = 300)\n",
    "\n",
    "# Create plot\n",
    "kp <- plotKaryotype(genome = \"hg38\", \n",
    "                    chromosomes = chromosomes_to_plot,\n",
    "                    plot.type = 1,\n",
    "                    cex = 0.7)\n",
    "\n",
    "kpPoints(kp, data = gr_strs, \n",
    "         y = 0.5,\n",
    "         col = \"darkblue\",\n",
    "         cex = 0.3,\n",
    "         pch = 16)\n",
    "\n",
    "#kpAddMainTitle(kp, \"Distribution of STRs in the human genome\", cex = 1.2)\n",
    "\n",
    "# Close device and save\n",
    "dev.off()\n",
    "\n",
    "cat(\"Plot saved to: results/karyotype_STRs_distribution.png\\n\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "563a2620",
   "metadata": {},
   "source": [
    "Heatmap showing STR density per 1 MB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "bf8347d1",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>agg_record_2147188300:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{agg\\textbackslash{}\\_record\\textbackslash{}\\_2147188300:} 2"
      ],
      "text/markdown": [
       "**agg_record_2147188300:** 2"
      ],
      "text/plain": [
       "agg_record_2147188300 \n",
       "                    2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Plot saved to: results/density_STRs.png\n"
     ]
    }
   ],
   "source": [
    "# Define standard somatic and sex chromosomes\n",
    "chromosomes_to_plot <- c(paste0(\"chr\", 1:22), \"chrX\", \"chrY\")\n",
    "\n",
    "# Bin parameters\n",
    "bin_size <- 1e6\n",
    "chr_lengths <- seqlengths(BSgenome.Hsapiens.UCSC.hg38)\n",
    "\n",
    "# Create bins for chromosomes of interest\n",
    "bins_list <- list()\n",
    "for (chr in chromosomes_to_plot) {\n",
    "  if (chr %in% names(chr_lengths)) {\n",
    "    chr_len <- chr_lengths[chr]\n",
    "    bins_df <- data.frame(\n",
    "      chr = chr,\n",
    "      start = seq(1, chr_len, by = bin_size),\n",
    "      end = pmin(seq(1, chr_len, by = bin_size) + bin_size - 1, chr_len)\n",
    "    )\n",
    "    bins_list[[chr]] <- bins_df\n",
    "  }\n",
    "}\n",
    "\n",
    "bins <- do.call(rbind, bins_list)\n",
    "bins_gr <- GRanges(seqnames = bins$chr,\n",
    "                   ranges = IRanges(start = bins$start, end = bins$end))\n",
    "\n",
    "# Calculate the density of STRs in bins using your complete GRanges\n",
    "overlaps <- countOverlaps(bins_gr, gr_strs)\n",
    "mcols(bins_gr)$density <- overlaps\n",
    "\n",
    "# Set color palette for heatmap\n",
    "colors <- c(\"white\", \"yellow\",\"#d94701\", \"#ee070bff\")\n",
    "\n",
    "# Set intervals for quantitative legend\n",
    "breaks <- quantile(mcols(bins_gr)$density, probs = seq(0, 1, length.out = length(colors) + 1))\n",
    "legend_labels <- paste0(round(breaks[-length(breaks)]), \" - \", round(breaks[-1]))\n",
    "\n",
    "png(\"results/density_STRs.png\", \n",
    "   width = 24, height = 14, units = \"in\", res = 600)\n",
    "\n",
    "\n",
    "# Create karyotype plot\n",
    "kp <- plotKaryotype(genome = \"hg38\",\n",
    "                    chromosomes = chromosomes_to_plot,\n",
    "                    plot.type = 1,\n",
    "                    cex = 1.2)\n",
    "\n",
    "# Plot density heatmap\n",
    "kpHeatmap(kp, data = bins_gr, y = bins_gr$density, colors = colors)\n",
    "\n",
    "# Add quantitative legend to chart\n",
    "legend(x = \"bottomright\",\n",
    "       legend = legend_labels,\n",
    "       fill = colors,\n",
    "       title = \"STRs density (per 1 MB)\",\n",
    "       bg = \"white\",\n",
    "       cex = 1.5)\n",
    "\n",
    "# Close the graphics device to save the file\n",
    "dev.off()\n",
    "\n",
    "cat(\"Plot saved to: results/density_STRs.png\\n\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
